#!/usr/bin/env nextflow
/*
========================================================================================
                         nf-core/dualrnaseq
========================================================================================
 nf-core/dualrnaseq Analysis Pipeline.
 #### Homepage / Documentation
 https://github.com/nf-core/dualrnaseq
----------------------------------------------------------------------------------------
*/

def helpMessage() {
    // TODO nf-core: Add to this help message with new command line parameters
    log.info nfcoreHeader()
    log.info"""

    Usage:

    The typical command for running the pipeline is as follows:

    nextflow run nf-core/dualrnaseq --reads '*_R{1,2}.fastq.gz' -profile docker

    Mandatory arguments:
      --reads [file]                Path to input data (must be surrounded with quotes)
      -profile [str]                Configuration profile to use. Can use multiple (comma separated)
                                    Available: conda, docker, singularity, test, awsbatch, <institute> and more

    Options:
      --genome [str]                  Name of iGenomes reference
      --single_end [bool]             Specifies that the input is single-end reads

    References                        If not specified in the configuration file or you wish to overwrite any of the references
      --fasta [file]                  Path to fasta reference

    Trimming:
      --a [str]                           adapter sequence for single-end reads or first reads of paired-end data
      --A [str]                           adapter sequence for second reads of paired-end data
      --quality-cutoff              cutoff to remove low-quality ends of reads. A single cutoff value is used to trim the 3’ end of reads. If two comma-separated cutoffs defined, the first value reprerents 5’ cutoff, and the second value defines 3’ cutoff.
      --skipTrimming                Skip trimming step


    Other options:
      --outdir [file]                 The output directory where the results will be saved
      --email [email]                 Set this parameter to your e-mail address to get a summary e-mail with details of the run sent to you when the workflow exits
      --email_on_fail [email]         Same as --email, except only send mail if the workflow is not successful
      --max_multiqc_email_size [str]  Theshold size for MultiQC report to be attached in notification email. If file generated by pipeline exceeds the threshold, it will not be attached (Default: 25MB)
      -name [str]                     Name for the pipeline run. If not specified, Nextflow will automatically generate a random mnemonic

    AWSBatch options:
      --awsqueue [str]                The AWSBatch JobQueue that needs to be set when running on AWSBatch
      --awsregion [str]               The AWS Region for your AWS Batch job to run on
      --awscli [str]                  Path to the AWS CLI tool
    """.stripIndent()
}

// Show help message
if (params.help) {
    helpMessage()
    exit 0
}

/*
 * SET UP CONFIGURATION VARIABLES
 */

// Check if genome exists in the config file
if (params.genomes && params.genome && !params.genomes.containsKey(params.genome)) {
    exit 1, "The provided genome '${params.genome}' is not available in the iGenomes file. Currently the available genomes are ${params.genomes.keySet().join(", ")}"
}


// reference genomes

params.fasta_host = params.genome_host ? params.genomes[ params.genome_host ].fasta_host ?: false : false
if (params.fasta_host) { ch_fasta_host = file(params.fasta_host, checkIfExists: true) }

params.fasta_pathogen = params.genome_pathogen ? params.genomes[ params.genome_pathogen ].fasta_pathogen ?: false : false
if (params.fasta_pathogen) { ch_fasta_pathogen = file(params.fasta_pathogen, checkIfExists: true) }

params.gff_host_tRNA = params.genome_host ? params.genomes[ params.genome_host ].gff_host_tRNA ?: false : false
if (params.gff_host_tRNA) { ch_gff_host_tRNA = file(params.gff_host_tRNA, checkIfExists: true) }

params.gff_host_genome = params.genome_host ? params.genomes[ params.genome_host ].gff_host ?: false : false
if (params.gff_host_genome) { ch_gff_host_genome = file(params.gff_host_genome, checkIfExists: true) }

params.gff_pathogen = params.genome_pathogen ? params.genomes[ params.genome_pathogen ].gff_pathogen ?: false : false
if (params.gff_pathogen) { ch_gff_pathogen = file(params.gff_pathogen, checkIfExists: true) }


if(params.read_transcriptome_fasta_host_from_file){
    params.transcriptome_host = params.genome_host ? params.genomes[ params.genome_host ].transcriptome_host ?: false : false
    if (params.transcriptome_host) { ch_transcriptome_host = file(params.transcriptome_host, checkIfExists: true) }
}

if(params.read_transcriptome_fasta_pathogen_from_file){
    params.transcriptome_pathogen = params.genome_pathogen ? params.genomes[ params.genome_pathogen ].transcriptome_pathogen ?: false : false
    if (params.transcriptome_pathogen) { ch_transcriptome_pathogen = file(params.transcriptome_pathogen, checkIfExists: true) }
}



 // parameters

if (params.run_salmon_selective_alignment){
	if (!params.libtype){
    exit 1, "Salmon: Please specify --libtype"
} else if (params.single_end & (params.libtype != 'U' && params.libtype != 'SR' && params.libtype != 'SF')) {
	    exit 1, "Salmon: Invalid library type --libtype ${params.libtype}! Library types available for single-end reads are:'U', 'SR', 'SF'."
} else if (!params.single_end & (params.libtype != 'IU' && params.libtype != 'ISR' && params.libtype != 'ISF' && params.libtype != 'MU' && params.libtype != 'MSR' && params.libtype != 'MSF' && params.libtype != 'OU' && params.libtype != 'OSR' && params.libtype != 'OSF' )) {
	    exit 1, "Salmon: Invalid library type --libtype ${params.libtype}! Library types available for paired-end reads are: 'IU', 'ISR', 'ISF', 'MU', 'MSR', 'MSF','OU', 'OSR', 'OSF'."
} 
}


if(params.mapping_statistics) {
	if (params.RNA_classes_to_replace_host) { ch_RNA_classes = file(params.RNA_classes_to_replace_host, checkIfExists: true) }
}


// Has the run name been specified by the user?
//  this has the bonus effect of catching both -name and --name
custom_runName = params.name
if (!(workflow.runName ==~ /[a-z]+_[a-z]+/)) {
    custom_runName = workflow.runName
}

if (workflow.profile.contains('awsbatch')) {
    // AWSBatch sanity checking
    if (!params.awsqueue || !params.awsregion) exit 1, "Specify correct --awsqueue and --awsregion parameters on AWSBatch!"
    // Check outdir paths to be S3 buckets if running on AWSBatch
    // related: https://github.com/nextflow-io/nextflow/issues/813
    if (!params.outdir.startsWith('s3:')) exit 1, "Outdir not on S3 - specify S3 Bucket to run on AWSBatch!"
    // Prevent trace files to be stored on S3 since S3 does not support rolling files.
    if (params.tracedir.startsWith('s3:')) exit 1, "Specify a local tracedir or run without trace! S3 cannot be used for tracefiles."
}

// Stage config files
ch_multiqc_config = file("$baseDir/assets/multiqc_config.yaml", checkIfExists: true)
ch_multiqc_custom_config = params.multiqc_config ? Channel.fromPath(params.multiqc_config, checkIfExists: true) : Channel.empty()
ch_output_docs = file("$baseDir/docs/output.md", checkIfExists: true)





/*
 * Create a channel for input read files
 */
if (params.readPaths) {
    if (params.single_end) {
        Channel
            .from(params.readPaths)
            .map { row -> [ row[0], [ file(row[1][0], checkIfExists: true) ] ] }
            .ifEmpty { exit 1, "params.readPaths was empty - no input files supplied" }
            .into { ch_read_files_fastqc; trimming_reads; raw_read_count }
    } else {
        Channel
            .from(params.readPaths)
            .map { row -> [ row[0], [ file(row[1][0], checkIfExists: true), file(row[1][1], checkIfExists: true) ] ] }
            .ifEmpty { exit 1, "params.readPaths was empty - no input files supplied" }
            .into { ch_read_files_fastqc; trimming_reads; raw_read_count }
    }
} else {
    Channel
        .fromFilePairs(params.reads, size: params.single_end ? 1 : 2)
        .ifEmpty { exit 1, "Cannot find any reads matching: ${params.reads}\nNB: Path needs to be enclosed in quotes!\nIf this is single-end data, please specify --single_end on the command line." }
        .into { ch_read_files_fastqc; trimming_reads; raw_read_count}
}





Channel
    .value( ch_fasta_pathogen)
    .collect()
    .into { genome_fasta_pathogen_to_combine; genome_fasta_pathogen_ref_names; genome_fasta_pathogen_to_transcriptome}

Channel
    .value( ch_fasta_host )
    .into { genome_fasta_host_to_combine; genome_fasta_host_ref_names; genome_fasta_host_to_transcriptome; genome_fasta_host_to_transcriptome_tRNA}



if(params.gff_host_tRNA){
	Channel
	    .value(ch_gff_host_tRNA)
	    .into {change_attrubute_gff_host_tRNA_salmon_alignment;gff_host_create_transcriptome_tRNA; combine_gff_host_tRNA_htseq}

	Channel
	    .value(ch_gff_host_genome)
	    .into { combine_gff_host_genome_star_salmon;gff_host_create_transcriptome; combine_gff_host_genome_htseq}
}else{
	Channel
	    .value(ch_gff_host_genome)
	    .into { gff_host_genome_salmon_alignment;gff_host_create_transcriptome; gff_host_genome_htseq; extract_annotations_host_gff_htseq}
}

Channel
    .value(ch_gff_pathogen)
    .into {gff_feature_quant_pathogen_salmon_alignment; gff_pathogen_create_transcriptome; gff_feature_quant_pathogen_htseq; extract_annotations_pathogen_gff_htseq}


if(params.read_transcriptome_fasta_host_from_file){
Channel
    .value(ch_transcriptome_host)
    .set {host_transcriptome_to_combine}
}


if(params.read_transcriptome_fasta_pathogen_from_file){
Channel
    .value(ch_transcriptome_pathogen)
    .set {pathogen_transcriptome_to_combine}
}



if (!params.skipTrimming){
	if(params.single_end){
		Channel
		    .value( params.a )
		    .set { adapter_sequence_3 }
	} else {
		Channel
		    .from (params.a, params.A)
		    .collect()
		    .set { adapter_sequence_3 }
	}
	Channel
	 .value(params.quality_cutoff)
	 .set { quality_cutoff}
}



if (params.run_salmon_selective_alignment | params.run_salmon_alignment_based_mode) {
	Channel
	    .value(params.gene_attribute_gff_to_create_transcriptome_host)
	    .into {host_gff_attribute_salmon_alignment; gene_attribute_gff_to_create_transcriptome_host_salmon; host_atr_collect_data_salmon; combine_annot_quant_pathogen; combine_annot_quant_host; atr_scatter_plot_pathogen; atr_scatter_plot_host; attribute_quant_stats_salmon; host_annotations_RNA_class_stats_pathogen; attribute_host_RNA_class_stats; host_atr_collect_data_salmon_alignment_mode; combine_annot_quant_pathogen_salmon_alignment_based; combine_annot_quant_host_salmon_alignment_based; atr_scatter_plot_pathogen_alignment; atr_scatter_plot_host_alignment; attribute_quant_stats_salmon_alignment;host_annotations_RNA_class_stats_pathogen_alignment; attribute_host_RNA_class_stats_alignment}

	Channel
	    .value(params.gene_feature_gff_to_create_transcriptome_host)
	    .collect()
	    .into { gene_feature_gff_host_salmon_alignment; gene_feature_gff_to_create_transcriptome_host_salmon}

	Channel
	    .value(params.gene_attribute_gff_to_create_transcriptome_pathogen)
	    .into {pathogen_gff_attribute_salmon_alignment; gene_attribute_gff_to_create_transcriptome_pathogen_salmon}

	Channel
	    .value(params.gene_feature_gff_to_create_transcriptome_pathogen)
	    .collect()
	    .into {gene_feature_to_quantify_pathogen_salmon_alignment; gene_feature_to_extract_annotations_pathogen; gene_feature_gff_to_create_transcriptome_pathogen_salmon}

}


if (params.run_salmon_selective_alignment){

	Channel
	    .value(params.kmer_length)
	    .set {kmer_length_salmon_index}


	Channel
	    .value(params.libtype)
	    .into {libtype_salmon; libtype_salmon_alignment_mode}

}


if(params.run_htseq_uniquely_mapped | params.run_htseq_multi_mapped){

	Channel
	    .value(params.gene_feature_gff_to_quantify_host)
	    .collect()
	    .into {gene_feature_to_quantify_host; gene_feature_to_extract_annotations_host_htseq}

	Channel
	    .value(params.gene_feature_gff_to_quantify_pathogen)
	    .collect()
	    .into {gene_feature_to_quantify_pathogen; gene_feature_to_extract_annotations_pathongen_htseq}

	Channel
	    .value(params.pathogen_gff_attribute)
	    .into { pathogen_gff_attribute; pathogen_gff_attribute_to_extract_annotations_htseq}

	Channel
	    .value(params.host_gff_attribute)
	    .into { host_gff_attribute_to_pathogen; host_gff_attribute_htseq; host_gff_attribute_htseq_combine; host_gff_attribute_to_extract_annotations_htseq; host_gff_attribute_mapping_stats_htseq; host_gff_attribute_RNA_class_pathogen_htseq; host_gff_attribute_RNA_class_host_htseq; combine_annot_quant_pathogen_host_gff_attribute}
}



if(params.mapping_statistics) {
Channel
    .value(ch_RNA_classes)
    .into { RNA_classes_to_replace; RNA_classes_to_replace_alignment; RNA_classes_to_replace_htseq_uniquely_mapped}
}





// Header log info
log.info nfcoreHeader()
def summary = [:]
if (workflow.revision) summary['Pipeline Release'] = workflow.revision
summary['Run Name']         = custom_runName ?: workflow.runName





// TODO nf-core: Report custom parameters here
summary['Reads']            = params.reads
summary['Host fasta Ref']        = params.fasta_host
summary['Pathogen fasta Ref']        = params.fasta_pathogen
summary['Host tRNA gff Ref']        = params.gff_host_tRNA
summary['Host genome gff Ref']        = params.gff_host_genome
summary['Pathogen genome gff Ref']        = params.gff_pathogen
summary['Data Type']        = params.single_end ? 'Single-End' : 'Paired-End'
summary['Max Resources']    = "$params.max_memory memory, $params.max_cpus cpus, $params.max_time time per job"
if (workflow.containerEngine) summary['Container'] = "$workflow.containerEngine - $workflow.container"
summary['Output dir']       = params.outdir
summary['Launch dir']       = workflow.launchDir
summary['Working dir']      = workflow.workDir
summary['Script dir']       = workflow.projectDir
summary['User']             = workflow.userName
if (workflow.profile.contains('awsbatch')) {
    summary['AWS Region']   = params.awsregion
    summary['AWS Queue']    = params.awsqueue
    summary['AWS CLI']      = params.awscli
}
summary['Config Profile'] = workflow.profile
if (params.config_profile_description) summary['Config Description'] = params.config_profile_description
if (params.config_profile_contact)     summary['Config Contact']     = params.config_profile_contact
if (params.config_profile_url)         summary['Config URL']         = params.config_profile_url
if (params.email || params.email_on_fail) {
    summary['E-mail Address']    = params.email
    summary['E-mail on failure'] = params.email_on_fail
    summary['MultiQC maxsize']   = params.max_multiqc_email_size
}
log.info summary.collect { k,v -> "${k.padRight(18)}: $v" }.join("\n")
log.info "-\033[2m--------------------------------------------------\033[0m-"

// Check the hostnames against configured profiles
checkHostname()

Channel.from(summary.collect{ [it.key, it.value] })
    .map { k,v -> "<dt>$k</dt><dd><samp>${v ?: '<span style=\"color:#999999;\">N/A</a>'}</samp></dd>" }
    .reduce { a, b -> return [a, b].join("\n            ") }
    .map { x -> """
    id: 'nf-core-dualrnaseq-summary'
    description: " - this information is collected when the pipeline is started."
    section_name: 'nf-core/dualrnaseq Workflow Summary'
    section_href: 'https://github.com/nf-core/dualrnaseq'
    plot_type: 'html'
    data: |
        <dl class=\"dl-horizontal\">
            $x
        </dl>
    """.stripIndent() }
    .set { ch_workflow_summary }







/*
 * Parse software version numbers
 */
/*
process get_software_versions {
    publishDir "${params.outdir}/pipeline_info", mode: 'copy',
        saveAs: { filename ->
                      if (filename.indexOf(".csv") > 0) filename
                      else null
                }

    output:
    file 'software_versions_mqc.yaml' into ch_software_versions_yaml
    file "software_versions.csv"

    script:
    // TODO nf-core: Get all tools to print their version number here

    """
    echo $workflow.manifest.version > v_pipeline.txt
    echo $workflow.nextflow.version > v_nextflow.txt
    echo cutadapt --version > v_cutadapt.txt
    fastqc --version > v_fastqc.txt

    multiqc --version > v_multiqc.txt
    scrape_software_versions.py &> software_versions_mqc.yaml
    """
}

*/

/*
 *  create chimeric reference files
 */



/*
 * combine pathogen and host genome fasta files
 */

process combine_pathogen_host_fasta_genome {
    tag "combine_genome_fasta_files"
    publishDir "${params.outdir}/references", mode: 'copy'
    storeDir "${params.outdir}/references"

    label 'process_high'

    input:
    file(host_fa) from genome_fasta_host_to_combine
    file(pathogen_fa) from genome_fasta_pathogen_to_combine 

    output:
    file "host_pathogen.fasta" into host_pathogen_fasta_index
    file "host_pathogen.fasta" into genome_fasta_file_host_pathogen_to_decoy_transcriptome

    script:
    """
    cat $pathogen_fa $host_fa > host_pathogen.fasta
    """
}





/*
 * chimeric gff file
 */

if(params.run_htseq_uniquely_mapped| params.run_htseq_multi_mapped ) {

	/*
	 * combine host genome and tRNA gff files
	 */

	if(params.gff_host_tRNA){
		process combine_host_genome_tRNA_gff_files_htseq {
		    publishDir "${params.outdir}/references", mode: 'copy'
		    storeDir "${params.outdir}/references"
		    tag "combine_host_genome_tRNA_gff_files_htseq"
		    
		    label 'process_high'

		    input:
		    file(host_gff_genome) from combine_gff_host_genome_htseq
		    file(host_gff_tRNA) from combine_gff_host_tRNA_htseq

		    output:
		    file "${outfile_name}" into gff_host_genome_htseq
		    file "${outfile_name}" into split_tab_host_genome_gff_htseq
//		    file "${outfile_name}" into split_tab_host_genome_gff_htseq_m_m
		    file "${outfile_name}" into extract_annotations_host_gff_htseq

		    script:
		    outfile_name = host_gff_genome[0].toString().replaceAll(/.gff3|.gff/,"_with_tRNA.gff3")
		    """
		    cat $host_gff_genome $host_gff_tRNA > ${outfile_name}
		    """
		}
	}


	process replace_gene_feature_gff_host_htseq {
	    publishDir "${params.outdir}/references", mode: 'copy'
	    storeDir "${params.outdir}/references"
	    tag "replace_gene_feature_gff_host_htseq"

	    label 'process_high'

	    input:
	    file(gff) from gff_host_genome_htseq
	    val(features) from gene_feature_to_quantify_host

	    output:
	    file "${outfile_name}" into combine_gff_host 

	    script:
	    outfile_name = gff[0].toString().replaceAll(/.gff3|.gff/,"_quant_feature.gff3")
	    """
	    $workflow.projectDir/bin/replace_feature_gff.sh $gff ${outfile_name} $features
	    """
	}



	process replace_gene_feature_gff_pathogen_htseq {
	    publishDir "${params.outdir}/references", mode: 'copy'
	    storeDir "${params.outdir}/references"
	    tag "replace_gene_feature_gff_pathogen_htseq"

	    label 'process_high'

	    input:
	    file(gff) from gff_feature_quant_pathogen_htseq
	    val(features) from gene_feature_to_quantify_pathogen

	    output:
	    file "${outfile_name}" into to_replace_gff_attribute

	    script:
	    outfile_name = gff[0].toString().replaceAll(/.gff3|.gff/,"_quant_feature.gff3")
	    """
	    $workflow.projectDir/bin/replace_feature_gff.sh $gff ${outfile_name} $features
	    """
	}


	process replace_attribute_pathogen_gff_htseq {

	    publishDir "${params.outdir}/references", mode: 'copy'
	    storeDir "${params.outdir}/references"
	    tag "replace_attribute_pathogen_gff_htseq"

	    label 'process_high'

	    input:
	    file(gff) from to_replace_gff_attribute
	    val(host_attribute) from host_gff_attribute_to_pathogen
	    val(pathogen_attribute) from pathogen_gff_attribute

	    output:
	    file "${outfile_name}" into combine_gff_pathogen

	    script:
	    outfile_name = gff[0].toString().replaceAll(/.gff3|.gff/,"_new_attribute.gff3")
	    """
	    $workflow.projectDir/bin/replace_pathogen_attribute_gff.sh $gff ${outfile_name} $host_attribute $pathogen_attribute
	    """
	}


	/*
	 * combine pathogen and host genome gff files
	 */

	process combine_pathogen_host_gff_files_htseq {
	    publishDir "${params.outdir}/references", mode: 'copy' 
	    storeDir "${params.outdir}/references"
	    tag "combine_pathogen_host_gff_files_htseq"

	    label 'process_high'
	 
	    input:
	    file(host_gff) from combine_gff_host
	    file(pathogen_gff_genome) from combine_gff_pathogen

	    output:
	    file "host_pathogen_htseq.gff" into quantification_gff_u_m
//	    file "host_pathogen_htseq.gff" into quantification_gtf_u_m
	    file "host_pathogen_htseq.gff" into gff_host_pathogen_star_htseq_alignment_gff

	    script:
	    """
	    cat $pathogen_gff_genome $host_gff > host_pathogen_htseq.gff
	    """
	}


	/*
	* extract annotations from gff files - for RNA class statistics + to merge with quantification results
	*/

	process extract_annotations_pathogen_htseq {
	    publishDir "${params.outdir}/references", mode: 'copy'
	    storeDir "${params.outdir}/references"
	    tag "extract_annotations_pathogen_htseq"

	    label 'main_env'
	    label 'process_high'

	    input:
	    file gff from extract_annotations_pathogen_gff_htseq
	    val(features) from gene_feature_to_extract_annotations_pathongen_htseq
	    val(pathogen_attribute) from pathogen_gff_attribute_to_extract_annotations_htseq

	    output:
	    file "${outfile_name}*_htseq.csv" into pathogen_annotations_RNA_class_stats_htseq
	    file "${outfile_name}*_htseq.csv" into annotation_pathogen_combine_quant_htseq_u_m
	    file "${outfile_name}*_htseq.csv" into annotation_pathogen_split_quant_htseq

	    script:
	    outfile_name = gff[0].toString().replaceAll(/.gff3|.gff/,"")
	    """
	    python $workflow.projectDir/bin/extract_annotations_from_gff.py -gff $gff -f $features -a $pathogen_attribute -org pathogen -q_tool htseq -o ${outfile_name}
	    """

	}


	process extract_annotations_host_htseq {
	    publishDir "${params.outdir}/references", mode: 'copy'
	    storeDir "${params.outdir}/references"
	    tag "extract_annotations_host_htseq"

	    label 'main_env'
	    label 'process_high'

	    input:
	    file gff from extract_annotations_host_gff_htseq
	    val(features) from gene_feature_to_extract_annotations_host_htseq
	    val(host_attribute) from host_gff_attribute_to_extract_annotations_htseq

	    output:
	    file "${outfile_name}*_htseq.csv" into host_annotations_RNA_class_stats_htseq
//	    file "${outfile_name}*_htseq.csv" into annotation_host_combine_quant_htseq
	    file "${outfile_name}*_htseq.csv" into annotation_host_split_quant_htseq

	    script:
	    outfile_name = gff[0].toString().replaceAll(/.gff3|.gff/,"")
	    """
	    python $workflow.projectDir/bin/extract_annotations_from_gff.py -gff $gff -f $features -a $host_attribute -org host -q_tool htseq -o ${outfile_name}
	    """
	}


}





if(params.run_star) {

	if(params.mapping_statistics) {
		/*
		* extract reference names from genome fasta files - mapping stats
		*/

		process extract_reference_names_host_star {
		    publishDir "${params.outdir}/references", mode: 'copy' 
		    storeDir "${params.outdir}/references"
		    tag "extract_reference_names_host_star"

		    label 'process_high'

		    input:
		    file(host_fa) from genome_fasta_host_ref_names

		    output:
		    file "reference_host_names.txt" into reference_host_names_uniquelymapped 
		    file "reference_host_names.txt" into reference_host_names_crossmapped_find
		    file "reference_host_names.txt" into reference_host_names_multimapped

		    script:
		    """
		    $workflow.projectDir/bin/extract_reference_names_from_fasta_files.sh reference_host_names.txt $host_fa
		    """
		}


		process extract_reference_names_pathogen_star {
		    publishDir "${params.outdir}/references", mode: 'copy' 
		    storeDir "${params.outdir}/references"
		    tag "extract_reference_names_pathogen_star"

		    label 'process_high'

		    input:
		    file(pathogen_fa) from genome_fasta_pathogen_ref_names

		    output:
		    file "reference_pathogen_names.txt" into reference_pathogen_names_uniquelymapped
		    file "reference_pathogen_names.txt" into reference_pathogen_crossmapped_find
		    file "reference_pathogen_names.txt" into reference_pathogen_names_multimapped

		    script:
		    """
		    $workflow.projectDir/bin/extract_reference_names_from_fasta_files.sh reference_pathogen_names.txt $pathogen_fa
		    """
		}
	}
}








if(params.run_salmon_selective_alignment | params.run_salmon_alignment_based_mode) {
	if(params.gff_host_tRNA){
		process replace_attribute_host_tRNA_gff_star_salmon {
		    tag "replace_attribute_host_tRNA_gff"
		    publishDir "${params.outdir}/references", mode: 'copy'
		    storeDir "${params.outdir}/references"

		    label 'process_high'

		    input:
		    file(gff) from change_attrubute_gff_host_tRNA_salmon_alignment
		    val(host_attribute) from host_gff_attribute_salmon_alignment

		    output:
		    file "${outfile_name}" into combine_host_gffs

		    script:
		    outfile_name = gff[0].toString().replaceAll(/.gff3|.gff/,"_Parent_attribute.gff3")
		    """
		    $workflow.projectDir/bin/replace_pathogen_attribute_gff.sh $gff ${outfile_name} Parent $host_attribute
		    """
		}

		process combine_host_genome_tRNA_gff_star_salmon {
		    tag "combine_host_genome_tRNA_gff"
		    publishDir "${params.outdir}/references", mode: 'copy'
		    storeDir "${params.outdir}/references"
		    
		    label 'process_high'

		    input:
		    file(host_gff_genome) from combine_gff_host_genome_star_salmon
		    file(host_gff_tRNA) from combine_host_gffs

		    output:
		    file "${outfile_name}" into gff_host_genome_salmon_alignment

		    script:
		    outfile_name = host_gff_genome[0].toString().replaceAll(/.gff3|.gff/,"_with_tRNA_STAR_salmon.gff3")
		    """
		    cat $host_gff_genome $host_gff_tRNA > ${outfile_name}
		    """
		}
	}

	process replace_gene_feature_gff_host_salmon {
	    tag "replace_gene_feature_gff_host"
	    publishDir "${params.outdir}/references", mode: 'copy'
	    storeDir "${params.outdir}/references"

	    label 'process_high'

	    input:
	    file(gff) from gff_host_genome_salmon_alignment
	    val(features) from gene_feature_gff_host_salmon_alignment

	    output:
	    file "${outfile_name}" into combine_gff_host_salmon_alignment
	    file "${outfile_name}" into extract_annotations_host_gff_salmon

	    script:
	    outfile_name = gff[0].toString().replaceAll(/.gff3|.gff/,"_quant_feature_salmon_alignment.gff3")
	    """
	    $workflow.projectDir/bin/replace_feature_gff.sh $gff ${outfile_name} $features
	    """
	}


	process replace_attribute_pathogen_gff_star_salmon {
	    tag "replace_attribute_pathogen_gff"
	    publishDir "${params.outdir}/references", mode: 'copy'
	    storeDir "${params.outdir}/references"

	    label 'process_high'

	    input:
	    file(gff) from gff_feature_quant_pathogen_salmon_alignment
	    val(pathogen_attribute) from pathogen_gff_attribute_salmon_alignment

	    output:
	    file "${outfile_name}" into to_replace_gff_feature_salmon_alignment
	    file "${outfile_name}" into extract_annotations_pathogen_gff_salmon


	    script:
	    outfile_name = gff[0].toString().replaceAll(/.gff3|.gff/,"_Parent_attribute.gff3")
	    """
	    $workflow.projectDir/bin/replace_pathogen_attribute_gff.sh $gff ${outfile_name} Parent $pathogen_attribute
	    """
	}
}


if(params.run_salmon_alignment_based_mode){
	process replace_gene_feature_gff_pathogen_salmon {
	    tag "replace_gene_feature_gff_pathogen"
	    publishDir "${params.outdir}/references", mode: 'copy'
	    storeDir "${params.outdir}/references"

	    label 'process_high'

	    input:
	    file(gff) from to_replace_gff_feature_salmon_alignment
	    val(features) from gene_feature_to_quantify_pathogen_salmon_alignment

	    output:
	    file "${outfile_name}" into combine_gff_pathogen_salmon_alignment

	    script:
	    outfile_name = gff[0].toString().replaceAll(/.gff3|.gff/,"_quant_feature_salmon_alignment.gff3")
	    """
	    $workflow.projectDir/bin/replace_feature_gff.sh $gff ${outfile_name} $features
	    """
	}

	process combine_pathogen_host_gff_files_salmon {
	    tag "combine_pathogen_host_gff"
	    publishDir "${params.outdir}/references", mode: 'copy' 
	    storeDir "${params.outdir}/references"

	    label 'process_high'
	 
	    input:
	    file(host_gff) from combine_gff_host_salmon_alignment
	    file(pathogen_gff_genome) from combine_gff_pathogen_salmon_alignment

	    output:
	    file "host_pathogen_star_alignment_mode.gff" into genome_gff_star_index
	    file "host_pathogen_star_alignment_mode.gff" into gff_host_pathogen_star_alignment_gff
	    file "host_pathogen_star_alignment_mode.gff" into gff_host_pathogen_star_salmon_alignment_gff

	    script:
	    """
	    cat $pathogen_gff_genome $host_gff > host_pathogen_star_alignment_mode.gff
	    """
	}

}


/*
* SALMON - extract gff annotations and create chimeric transcriptome 
*/


if(params.run_salmon_selective_alignment | params.run_salmon_alignment_based_mode) {


	/*
	* extract annotations from gff file - for RNA class statistics + to merge with quantification results
	*/


	process extract_annotations_pathogen_salmon {
	    publishDir "${params.outdir}/references", mode: 'copy'
	    storeDir "${params.outdir}/references"
	    tag "extract_gff_annotations_pathogen"

	    label 'main_env'
	    label 'process_high'

	    input:
	    file gff from extract_annotations_pathogen_gff_salmon
	    val(features) from gene_feature_to_extract_annotations_pathogen

	    output:
	    file "${outfile_name}*_salmon.csv" into pathogen_annotations_RNA_class_stats
	    file "${outfile_name}*_salmon.csv" into pathogen_annotations_RNA_class_stats_salmon_alignment
	    file "${outfile_name}*_salmon.csv" into annotation_pathogen_combine_quant
	    file "${outfile_name}*_salmon.csv" into annotation_pathogen_combine_quant_salmon_alignment_based

	    script:
	    outfile_name = "pathogen_gff_annotations"
	    """
	    python $workflow.projectDir/bin/extract_annotations_from_gff.py -gff $gff -f $features -a Parent -org pathogen -q_tool salmon -o ${outfile_name}
	    """
	}


	process extract_annotations_host_salmon {
	    publishDir "${params.outdir}/references", mode: 'copy'
	    storeDir "${params.outdir}/references"
	    tag "extract_gff_annotations_host"

	    label 'main_env'
	    label 'process_high'

	    input:
	    file gff from extract_annotations_host_gff_salmon

	    output:
	    file "${outfile_name}*_salmon.csv" into host_annotations_RNA_class_stats
	    file "${outfile_name}*_salmon.csv" into host_annotations_RNA_class_stats_salmon_alignment

	    file "${outfile_name}*_salmon.csv" into annotation_host_combine_quant
	    file "${outfile_name}*_salmon.csv" into annotation_host_combine_quant_salmon_alignment_based

	    script:
	    outfile_name = "host_gff_annotations"
	    """
	    python $workflow.projectDir/bin/extract_annotations_from_gff.py -gff $gff -f quant -a Parent -org host -q_tool salmon -o ${outfile_name}
	    """
	}


	/*
	 * create transcriptome fasta file - host
	 */

	if(!params.read_transcriptome_fasta_host_from_file){
		process create_transcriptome_fasta_host {
		    tag "create_transcriptome_fasta_host"
		    publishDir "${params.outdir}/references", mode: 'copy'
		    storeDir "${params.outdir}/references"

	            label 'main_env'
		    label 'process_high'

		    input:
		    file(gff) from gff_host_create_transcriptome
		    file(host_fa) from genome_fasta_host_to_transcriptome

		    output:
		    file "${outfile_name}" into host_transcriptome
		    file "${outfile_name}" into transcriptome_host_to_split_table_salmon_without_tRNA
		    file "${outfile_name}" into transcriptome_host_to_split_table_salmon_alignment_without_tRNA
		    file "${outfile_name}" into transcriptome_host_to_split_q_table_salmon_without_tRNA
		    file "${outfile_name}" into transcriptome_host_to_split_q_table_salmon_alignment_based_without_tRNA
		    file "${outfile_name}" into host_transcriptome_to_combine_host

		    script:
		    outfile_name = gff[0].toString().replaceAll(/.gff3|.gff/,"_transcriptome.fasta")
		    """
		    gffread -w $outfile_name -g $host_fa $gff
		    """
	}
		if(params.gff_host_tRNA){
			process create_transcriptome_fasta_host_tRNA {
			    tag "create_transcriptome_fasta_tRNA_host"
			    publishDir "${params.outdir}/references", mode: 'copy'
			    storeDir "${params.outdir}/references"

	                    label 'main_env'
			    label 'process_high'

			    input:
			    file(gff) from gff_host_create_transcriptome_tRNA
			    file(host_fa) from genome_fasta_host_to_transcriptome_tRNA
			    val(features) from gene_feature_gff_to_create_transcriptome_host_salmon
			    val(attribute) from gene_attribute_gff_to_create_transcriptome_host_salmon

			    output:
			    file "${outfile_name}" into host_transcriptome_to_combine_tRNA

			    script:
			    outfile_name = gff[0].toString().replaceAll(/.gff3|.gff/,"_transcriptome.fasta")
			    """
			    python $workflow.projectDir/bin/gff_to_fasta_transcriptome.py -fasta $host_fa -gff $gff  -f $features -a $attribute -o $outfile_name
			    """
			}
			process combine_host_fasta_transcriptomes {
			    tag "combine_host_genome_tRNA_fasta_transcriptome"
			    publishDir "${params.outdir}/references", mode: 'copy'
			    storeDir "${params.outdir}/references"

			    label 'process_high'

			    input:
			    file(host_tr_fa) from host_transcriptome_to_combine_host
			    file(host_tRNA_tr_fa) from host_transcriptome_to_combine_tRNA

			    output:
			    file "host_transcriptome.fasta" into host_transcriptome_genome_tRNA
                            file "host_transcriptome.fasta" into transcriptome_host_to_split_table_salmon_with_tRNA
                            file "host_transcriptome.fasta" into transcriptome_host_to_split_table_salmon_alignment_with_tRNA
                            file "host_transcriptome.fasta" into transcriptome_host_to_split_q_table_salmon_with_tRNA
                            file "host_transcriptome.fasta" into transcriptome_host_to_split_q_table_salmon_alignment_based_with_tRNA

 			    script:
			    """
			    cat $host_tr_fa $host_tRNA_tr_fa > host_transcriptome.fasta
			    """
			}
		}

	host_transcriptome_to_combine = params.gff_host_tRNA ? host_transcriptome_genome_tRNA : host_transcriptome

	transcriptome_host_to_split_q_table_salmon = params.gff_host_tRNA ? transcriptome_host_to_split_q_table_salmon_with_tRNA : transcriptome_host_to_split_q_table_salmon_without_tRNA

	transcriptome_host_to_split_table_salmon = params.gff_host_tRNA ? transcriptome_host_to_split_table_salmon_with_tRNA : transcriptome_host_to_split_table_salmon_without_tRNA

	transcriptome_host_to_split_q_table_salmon_alignment_based = params.gff_host_tRNA ? transcriptome_host_to_split_q_table_salmon_alignment_based_with_tRNA : transcriptome_host_to_split_q_table_salmon_alignment_based_without_tRNA

	transcriptome_host_to_split_table_salmon_alignment = params.gff_host_tRNA ? transcriptome_host_to_split_table_salmon_alignment_with_tRNA : transcriptome_host_to_split_table_salmon_alignment_without_tRNA
	}




	/*
	 * create transcriptome fasta file - pathogen
	 */

	if(!params.read_transcriptome_fasta_pathogen_from_file){
		process create_transcriptome_fasta_pathogen {
		    tag "create_transcriptome_fasta_pathogen"
		    publishDir "${params.outdir}/references", mode: 'copy'
		    storeDir "${params.outdir}/references"

		    label 'main'
		    label 'process_high'

		    input:
		    file(gff) from gff_pathogen_create_transcriptome
		    file(pathogen_fa) from genome_fasta_pathogen_to_transcriptome
		    val(features) from gene_feature_gff_to_create_transcriptome_pathogen_salmon
		    val(attribute) from gene_attribute_gff_to_create_transcriptome_pathogen_salmon

		    output:
		    file "${outfile_name}" into pathogen_transcriptome_to_combine
		    file "${outfile_name}" into transcriptome_pathogen_to_split_table_salmon
		    file "${outfile_name}" into transcriptome_pathogen_to_split_table_salmon_alignment
		    file "${outfile_name}" into transcriptome_pathogen_to_split_q_table_salmon
		    file "${outfile_name}" into transcriptome_pathogen_to_split_q_table_salmon_alignment_based

		    script:
		    outfile_name = gff[0].toString().replaceAll(/.gff3|.gff/,"_transcriptome.fasta")
		    """
		    python $workflow.projectDir/bin/gff_to_fasta_transcriptome.py -fasta $pathogen_fa -gff $gff -f $features -a $attribute  -o $outfile_name
		    """
		}
	}


	/*
	 * combine pathogen and host transcriptome fasta files
	 */


	process combine_pathogen_host_fasta_transcriptome {
	    tag "combine_pathogen_host_fasta_transcriptome"
	    publishDir "${params.outdir}/references", mode: 'copy'
	    storeDir "${params.outdir}/references"

	    label 'process_high'

	    input:
	    file(host_tr_fa) from host_transcriptome_to_combine
	    file(pathogen_tr_fa) from pathogen_transcriptome_to_combine

	    output:
	    file "host_pathogen_transcriptome.fasta" into host_pathogen_transciptom_fasta_index
	    file "host_pathogen_transcriptome.fasta" into transcriptome_fasta_file_host_pathogen_to_decoy_transcriptome
	    file "host_pathogen_transcriptome.fasta" into transcriptome_salmon_alignment_based_mode

	    script:
	    """
	    cat $pathogen_tr_fa $host_tr_fa > host_pathogen_transcriptome.fasta
	    """
	}


}




/*
 * STEP 1 - FastQC
 */
process fastqc {
    tag "$name"
    label 'process_medium'
    label 'main_env'
    publishDir "${params.outdir}/fastqc", mode: 'copy',
        saveAs: { filename ->
                      filename.indexOf(".zip") > 0 ? "zips/$filename" : "$filename"
                }
    storeDir "${params.outdir}/fastqc"

    input:
    set val(name), file(reads) from ch_read_files_fastqc

    output:
    file "*_fastqc.{zip,html}" into ch_fastqc_results

    script:
    """
    fastqc --quiet --threads $task.cpus --noextract $reads
    """
}

/*
 *  STEP 2 - Trimming
 */

if (!params.skipTrimming) {
	process trimming {
	    tag "$name_reads"
	    publishDir "${params.outdir}/trimming", mode: 'copy'
	    storeDir "${params.outdir}/trimming"

	    label 'main_env'
	    label 'process_high'

	    input:
	    set val(name), file(reads) from trimming_reads
	    val adapter_seq_3 from adapter_sequence_3
	    val q_value from quality_cutoff

	    output:
	    set val(name_sample), file("${name_sample}{_1,_2,}_trimmed.fastq.gz") into trimming_results_star_htseq, trimming_results_to_salmon, trimming_results_to_qc, trimming_results_star_salmon
	    file("${name_sample}{_1,_2,}_trimmed.fastq.gz") into count_reads

	    script:
	if (params.single_end){
	    name_reads = reads.toString().replaceAll(/:/,"_")
	    name_reads = name_reads.replaceAll(/-/,"_")
	    name_out = name_reads.replaceAll(/.fastq.gz|.fq.gz|.fastq|.fq/,"_trimmed.fastq.gz")
	    name_sample = name_reads.replaceAll(/.fastq.gz|.fq.gz|.fastq|.fq/,"")
	    """
	    cutadapt -j ${task.cpus} -q $q_value -a $adapter_seq_3 -m 1 -o ${name_out} $reads 
	    """
	} else{
	    name_reads = name.replaceAll(/:/,"_")
	    name_reads = name_reads.replaceAll(/-/,"_")
	    name_sample = name_reads.replaceAll(/.fastq.gz|.fq.gz|.fastq|.fq/,"")
	    name_1 = reads[0][0].toString().replaceAll(/:/,"_")
	    name_1 = name_1.replaceAll(/-/,"_")
	    name_1 = name_1.replaceAll(/.fastq.gz|.fq.gz|.fastq|.fq/,"_trimmed.fastq.gz")
	    name_2 = reads[1][0].toString().replaceAll(/:/,"_")
	    name_2 = name_2.replaceAll(/-/,"_")
	    name_2 = name_2.replaceAll(/.fastq.gz|.fq.gz|.fastq|.fq/,"_trimmed.fastq.gz") 
	    """
	    cutadapt -j ${task.cpus} -q $q_value -a ${adapter_seq_3[0]} -A ${adapter_seq_3[1]} -o ${name_1} -p ${name_2} -m 1 ${reads[0]} ${reads[1]}
	    """
	}
	}
}else{
   trimming_reads
      .into {trimming_results_to_qc; trimming_results_star_htseq; trimming_results_to_salmon; count_reads; trimming_results_star_salmon}
}



/*
 * STEP 3 -FastQC after trimming 
 */

if (!params.skipTrimming) {
	process fastqc_after_trimming {
	    tag "$sample_name"
	    publishDir "${params.outdir}/fastqc_after_trimming", mode: 'copy', saveAs: {filename -> filename.indexOf(".zip") > 0 ? "zips/$filename" : "$filename"}
	    storeDir "${params.outdir}/fastqc_after_trimming"

	    label 'main_env'
	    label 'process_medium'

	    input:
	    set val(name),file(reads) from trimming_results_to_qc

	    output:
	    file "${sample_name}{_1,_2,}_trimmed_fastqc.{zip,html}" into raw_reads_fastqc_trim

	    script:
	    sample_name = name.replaceFirst(/.fastq.gz|.fq.gz|.fastq|.fq/, "")
	    """
	    fastqc --threads ${task.cpus} --quiet --noextract $reads
	    """
	}
}else{
   trimming_results_to_qc
         .set{raw_reads_fastqc_trim}
}



if(params.mapping_statistics) {

	/*
	* count total number of raw reads
	*/


	raw_read_count
		.map { tag, file -> file }
		.set {raw_read_count_file}


	process count_total_reads {
	    tag "count_total_reads"
	    publishDir "${params.outdir}/mapping_statistics", mode: 'copy'
	    storeDir "${params.outdir}/mapping_statistics"

	    label 'process_high'

	    input:
	    file(fastq) from raw_read_count_file.collect()

	    output:
	    file "total_raw_reads_fastq.csv" into collect_total_reads_raw_salmon
	    file "total_raw_reads_fastq.csv" into collect_total_reads_raw_salmon_alignment
	    file "total_raw_reads_fastq.csv" into collect_total_reads_raw_htseq_uniquely_mapped

	    script:
	    """
	    $workflow.projectDir/bin/count_total_reads.sh $fastq >> total_raw_reads_fastq.csv
	    """
	}



	/*
	* count total number of reads after trimming
	*/


	process count_total_trimmed_reads {
	    tag "count_total_trimmed_reads"
	    publishDir "${params.outdir}/mapping_statistics", mode: 'copy'
	    storeDir "${params.outdir}/mapping_statistics"

	    label 'process_high'

	    input:
	    file(fastq) from count_reads.collect()

	    output:
	    file "total_trimmed_reads_fastq.csv" into mapping_stats_total_processed_reads_alignment

	    script:
	    """
	    $workflow.projectDir/bin/count_total_reads.sh $fastq >> total_trimmed_reads_fastq.csv
	    """
	}

}




if(params.run_salmon_selective_alignment) {


	/*
	 * create genome as decoy sequence 
	 */

	process create_decoy_transcriptome_file {
	    tag "create_decoy"
	    publishDir "${params.outdir}/references", mode: 'copy'
	    storeDir "${params.outdir}/references"

	    label 'process_high'

	    input:
	    file(host_pathogen_genome_fasta) from genome_fasta_file_host_pathogen_to_decoy_transcriptome
	    file(host_pathogen_transcriptome_fasta) from transcriptome_fasta_file_host_pathogen_to_decoy_transcriptome

            output:
            file "gentrome.fasta" into salmon_index_gentrome
            file "decoys.txt" into salmon_index_decoys

	    shell:
	    '''
	    grep ">" !{host_pathogen_genome_fasta} | cut -d " " -f 1 > decoys.txt
	    sed -i -e 's/>//g' decoys.txt
	    cat !{host_pathogen_transcriptome_fasta} !{host_pathogen_genome_fasta} > gentrome.fasta
	    '''
	}



	/*
	 * salmon - build an index
	 */

	process salmon_index {
	    tag "salmon_index"
            storeDir "${params.outdir}/salmon"

            label 'salmon_env'
            label 'process_high'

	    input:
	    file(gentrome) from salmon_index_gentrome
	    file(decoys) from salmon_index_decoys
	    val(kmer_length) from kmer_length_salmon_index

	    output:
	    file "transcripts_index" into salmon_index

	    script:
	    """
	    salmon index -t $gentrome -i transcripts_index --decoys $decoys -k $kmer_length -p ${task.cpus}
	    """
	}



	/*
	 * salmon - quantification
	 */


	process salmon_quantification {
//	    publishDir "${params.outdir}/salmon", mode: 'copy'
	    storeDir "${params.outdir}/salmon"
	    tag "${sample}"

            label 'salmon_env'
            label 'process_high'

	    input:
	    file(index) from salmon_index.collect()
	    set val(sample), file(reads) from trimming_results_to_salmon
	    val(libtype) from libtype_salmon

	    output:
	    set val(sample_name), file("${sample_name}") into split_table
	    file("${sample_name}") into salmon_files_to_combine
	    file("${sample_name}") into multiqc_salmon_quant
	    set val(sample_name), file("${sample_name}") into collect_processed_read_counts
 
	    script:
	UnmappedNames = params.writeUnmappedNames ? "--writeUnmappedNames" : ''
	softclip = params.softclipOverhangs ? "--softclipOverhangs" : ''
	incompatPrior = params.incompatPrior
	dumpEq = params.dumpEq ? "--dumpEq" : ''
if (params.single_end){
	    sample_name = sample.replaceFirst(/.fastq.gz|.fq.gz|.fastq|.fq/, "")
	    writeMappings = params.writeMappings ? "--writeMappings=$sample_name/mapping.sam" : ''
	    """
	    salmon quant -p ${task.cpus} -i $index -l $libtype -r $reads $softclip --incompatPrior $incompatPrior $UnmappedNames --validateMappings $dumpEq $writeMappings -o $sample_name 
	    """
} else{
	    sample_name = sample.replaceFirst(/.fastq.gz|.fq.gz|.fastq|.fq/, "")
	    writeMappings = params.writeMappings ? "--writeMappings=$sample_name/mapping.sam" : ''
	    """
	    salmon quant -p ${task.cpus} -i $index -l $libtype -1 ${reads[0]} -2 ${reads[1]} $softclip --incompatPrior $incompatPrior $UnmappedNames --validateMappings $dumpEq $writeMappings -o $sample_name 
	    """
}
}


	/*
	 * salmon - split quantification tables into host and pathogen results
	 */


	process split_table_salmon_each {
            publishDir "${params.outdir}/salmon/${sample_name}", mode: 'copy'
            storeDir "${params.outdir}/salmon/${sample_name}"
            tag "split_quantification ${sample}"

            label 'main_env'
            label 'process_high'

            input:
            set val(sample_name), file ("salmon/*") from split_table
            file transcriptome_pathogen from transcriptome_pathogen_to_split_q_table_salmon
            file transcriptome_host from transcriptome_host_to_split_q_table_salmon

            output:
            set val(sample_name), file("host_quant.sf")
            set val(sample_name), file("pathogen_quant.sf")

            shell:
            '''
            grep ">" !{transcriptome_pathogen} | awk -F ">" '{ print $2 }' | awk 'NR==FNR{a[$0]=$0}NR>FNR{if($1==a[$1])print $0}' - salmon/*/quant.sf > pathogen_quantification_without_headers.sf
            awk 'NR==1 {print; exit}' salmon/*/quant.sf | cat - pathogen_quantification_without_headers.sf > pathogen_quant.sf

            grep ">" !{transcriptome_host} | awk -F ">" '{ print $2 }' | awk -F ' ' '{print $1}' | awk 'NR==FNR{a[$0]=$0}NR>FNR{if($1==a[$1])print $0}' - salmon/*/quant.sf > host_quantification_salmon_without_headers.sf
            awk 'NR==1 {print; exit}' salmon/*/quant.sf | cat - host_quantification_salmon_without_headers.sf > host_quant.sf
            '''
        }



	/*
	 * salmon - combine quantification results
	 */


	process combine_quantification_tables_salmon {
		    publishDir "${params.outdir}/salmon", mode: 'copy'
		    storeDir "${params.outdir}/salmon"
		    tag "combine_quantification_salmon"

		    label 'main_env'
		    label 'process_high'

		    input: 
		    file input_quantification from salmon_files_to_combine.collect()
		    val gene_attribute from host_atr_collect_data_salmon

		    output:
		    file "combined_quant.csv" into split_table_salmon 

		    script:
		    """
		    python $workflow.projectDir/bin/collect_quantification_data.py -i $input_quantification -q salmon -a $gene_attribute -org both
		    """
		}



	/*
	 * salmon - split quantification tables into host and pathogen results
	 */


	process split_quantification_tables_salmon {
	    publishDir "${params.outdir}/salmon", mode: 'copy'
	    storeDir "${params.outdir}/salmon"
            tag "split_quantification_table_salmon"

            label 'main_env'
            label 'process_high'

            input:
            file quant_table from split_table_salmon
            file transcriptome_pathogen from transcriptome_pathogen_to_split_table_salmon
            file transcriptome_host from transcriptome_host_to_split_table_salmon

            output:
            file 'host_quantification_salmon.csv' into host_quantification_mapping_stats_salmon 
            file 'pathogen_quantification_salmon.csv' into pathogen_quantification_mapping_stats_salmon
            file 'host_quantification_salmon.csv' into host_quantification_RNA_stats_salmon 
            file 'pathogen_quantification_salmon.csv' into pathogen_quantification_RNA_stats_salmon
            file 'host_quantification_salmon.csv' into quant_host_add_annotations 
            file 'pathogen_quantification_salmon.csv' into quant_pathogen_add_annotations
            file 'host_quantification_salmon.csv' into quant_scatter_plot_host
            file 'pathogen_quantification_salmon.csv' into quant_scatter_plot_pathogen

            shell:
            '''
            grep ">" !{transcriptome_pathogen} | awk -F ">" '{ print $2 }' | awk 'NR==FNR{a[$0]=$0}NR>FNR{if($1==a[$1])print $0}' - !{quant_table}  > pathogen_quantification_without_headers.csv
            awk 'NR==1 {print; exit}' !{quant_table} | cat - pathogen_quantification_without_headers.csv > pathogen_quantification_salmon.csv

            grep ">" !{transcriptome_host} | awk -F ">" '{ print $2 }' | awk -F ' ' '{print $1}' | awk 'NR==FNR{a[$0]=$0}NR>FNR{if($1==a[$1])print $0}' - !{quant_table}  > host_quantification_salmon_without_headers.csv
            awk 'NR==1 {print; exit}' !{quant_table} | cat - host_quantification_salmon_without_headers.csv > host_quantification_salmon.csv
            '''
        }



	/*
	 * salmon - 'combine annotations extracted from gff with quantification tables
	 */

	process combine_annotations_quant_pathogen {
		    publishDir "${params.outdir}/salmon", mode: 'copy'
		    storeDir "${params.outdir}/salmon"
		    tag "combine_annotations_quant_pathogen_salmon"

		    label 'main_env'
   		    label 'process_high'
		   
	            input: 
		    file quantification_table from quant_pathogen_add_annotations
		    file annotation_table from annotation_pathogen_combine_quant
		    val attribute from combine_annot_quant_pathogen

		    output:
		    file "pathogen_combined_quant_annotations.csv"

		    script:
		    """
		    $workflow.projectDir/bin/combine_quant_annotations.py -q $quantification_table -annotations $annotation_table -a $attribute -org pathogen
		    """
		}



	process combine_annotations_quant_host_salmon {
		    publishDir "${params.outdir}/salmon", mode: 'copy'
		    storeDir "${params.outdir}/salmon"
		    tag "combine_annotations_quant_host_salmon"

		    label 'main_env'
   		    label 'process_high'
		   
	            input: 
		    file quantification_table from quant_host_add_annotations
		    file annotation_table from annotation_host_combine_quant
		    val attribute from combine_annot_quant_host

		    output:
		    file "host_combined_quant_annotations.csv"

		    script:
		    """
		    $workflow.projectDir/bin/combine_quant_annotations.py -q $quantification_table -annotations $annotation_table -a $attribute -org host
		    """
		}



	if(params.mapping_statistics) {
		/*
		 * salmon - 'quantification_stats'
		 */

		process scatter_plot_pathogen_salmon {
		    publishDir "${params.outdir}/mapping_statistics/salmon/scatter_plots", mode: 'copy'
		    storeDir "${params.outdir}/mapping_statistics/salmon/scatter_plots"
		    tag "scatter_plot_salmon_pathogen"

		    label 'main_env'
		    label 'process_high'

		    input:
		    file quant_table from quant_scatter_plot_pathogen
		    val attribute from atr_scatter_plot_pathogen

		    output:
		    file ('*.pdf')

		    script:
		    """
		    python $workflow.projectDir/bin/scatter_plots.py -q $quant_table -a $attribute -org pathogen
		    """
		}



		process scatter_plot_host_salmon {
		    publishDir "${params.outdir}/mapping_statistics/salmon/scatter_plots", mode: 'copy'
		    storeDir "${params.outdir}/mapping_statistics/salmon/scatter_plots"
		    tag "scatter_plot_salmon_host"

		    label 'main_env'
		    label 'process_high'

		    input:
		    file quant_table from quant_scatter_plot_host
		    val attribute from atr_scatter_plot_host

		    output:
		    file ('*.pdf') 

		    script:
		    """
		    python $workflow.projectDir/bin/scatter_plots.py -q $quant_table -a $attribute -org host
		    """
		}


		process extract_processed_reads {
			    publishDir "${params.outdir}/mapping_statistics/salmon", mode: 'copy'
			    storeDir "${params.outdir}/mapping_statistics/salmon"
			    tag "extract_processed_reads"

			    label 'main_env'
	   		    label 'process_high'
			   
			    input: 
			    set val(sample_name), file ("salmon/*") from collect_processed_read_counts

			    output:
			    file "${sample_name}.txt" into collect_results

			    script:
			    """
			    $workflow.projectDir/bin/extract_processed_reads_salmon.sh salmon/*/aux_info/meta_info.json $sample_name
			    """
			}


		process collect_processed_reads {
			    publishDir "${params.outdir}/mapping_statistics/salmon", mode: 'copy'
			    storeDir "${params.outdir}/mapping_statistics/salmon"
			    tag "collect_processed_reads"

			    label 'main_env'
			    label 'process_high' 
			    
			    input: 
			    file process_reads from collect_results.collect()

			    output:
			    file "processed_reads_salmon.csv" into mapping_stats_total_reads

			    script:
			    """
			    cat $process_reads > processed_reads_salmon.csv
			    """
			}


		process salmon_quantification_stats {
		    publishDir "${params.outdir}/mapping_statistics/salmon", mode: 'copy'
		    storeDir "${params.outdir}/mapping_statistics/salmon"
		    tag "quantification_statistics salmon"

		    label 'main_env'
		    label 'process_high'

		    input:
		    file quant_table_host from host_quantification_mapping_stats_salmon
		    file quant_table_pathogen from pathogen_quantification_mapping_stats_salmon
		    val attribute from attribute_quant_stats_salmon
		    file total_processed_reads from mapping_stats_total_reads
		    file total_raw_reads from collect_total_reads_raw_salmon

		    output:
		    file ('salmon_host_pathogen_total_reads.csv') into salmon_mapped_stats_to_plot

		    script:
		    """
		    python $workflow.projectDir/bin/quantification_stats.py -q_p $quant_table_pathogen -q_h $quant_table_host -total_processed $total_processed_reads -total_raw $total_raw_reads -a $attribute -t salmon -o salmon_host_pathogen_total_reads.csv
		    """
		}


		process plot_salmon_mapping_stats_host_pathogen {
		    tag "$name2"
		    publishDir "${params.outdir}/mapping_statistics/salmon", mode: 'copy'
		    storeDir "${params.outdir}/mapping_statistics/salmon"

		    label 'main_env'
		    label 'process_high'

		    input:
		    file(stats) from salmon_mapped_stats_to_plot

		    output:
		    file "mapping_stats_*"

		    script:
		    """
		    python $workflow.projectDir/bin/plot_mapping_statistics.py -i $stats
		    """
		}



		/*
		 * salmon - RNA class statistics
		 */


		process RNA_class_statistics_salmon_pathogen {
		    publishDir "${params.outdir}/mapping_statistics/salmon/RNA_classes_pathogen", mode: 'copy'
		    storeDir "${params.outdir}/mapping_statistics/salmon/RNA_classes_pathogen"
		    tag "RNA class stistics"

		    label 'main_env'
		    label 'process_high'

		    input:
		    file quant_table from pathogen_quantification_RNA_stats_salmon
		    val attribute from host_annotations_RNA_class_stats_pathogen
		    file gene_annotations from pathogen_annotations_RNA_class_stats

		    output:
		    file "pathogen_RNA_classes_percentage.csv" into plot_RNA_stats_pathogen
		    file "pathogen_RNA_classes_percentage.csv" into plot_RNA_stats_pathogen_combined
		    file "pathogen_RNA_classes_sum_counts.csv"

		    script:
		    """
		    python $workflow.projectDir/bin/RNA_class_content.py -q $quant_table -a $attribute -annotations $gene_annotations -q_tool salmon -org pathogen
		    """
		}


		/*
		 * salmon - RNA class statistics
		 */


		process RNA_class_statistics_salmon_host {
		    publishDir "${params.outdir}/mapping_statistics/salmon/RNA_classes_host", mode: 'copy'
		    storeDir "${params.outdir}/mapping_statistics/salmon/RNA_classes_host"
		    tag "RNA class stistics"

		    label 'main_env'
		    label 'process_high'

		    input:
		    file quant_table from host_quantification_RNA_stats_salmon
		    val attribute from attribute_host_RNA_class_stats
		    file gene_annotations from host_annotations_RNA_class_stats
		    file rna_classes_to_replace from RNA_classes_to_replace

		    output:
		    file "host_RNA_classes_percentage.csv" into plot_RNA_stats_host
		    file "host_RNA_classes_percentage.csv" into plot_RNA_stats_host_combined
		    file "host_RNA_classes_sum_counts.csv"
		    file "host_gene_types_groups_*"

		    script:
		    """
		    python $workflow.projectDir/bin/RNA_class_content.py -q $quant_table -a $attribute -annotations $gene_annotations -rna $rna_classes_to_replace -q_tool salmon -org host
		    """
		}



		process plot_RNA_class_salmon_pathogen_each {
		    publishDir "${params.outdir}/mapping_statistics/salmon/RNA_classes_pathogen", mode: 'copy'
		    storeDir "${params.outdir}/mapping_statistics/salmon/RNA_classes_pathogen"
		    tag "plot RNA class stistics each"

		    label 'main_env'
		    label 'process_high'

		    input:
		    file stats_table from plot_RNA_stats_pathogen

		    output:
		    file "*.pdf"

		    script:
		    """
		    python $workflow.projectDir/bin/plot_RNA_class_stats_each.py -i $stats_table
		    """
		}

		process plot_RNA_class_salmon_host_each {
		    publishDir "${params.outdir}/mapping_statistics/salmon/RNA_classes_host", mode: 'copy'
		    storeDir "${params.outdir}/mapping_statistics/salmon/RNA_classes_host"
		    tag "plot RNA class stistics each"

		    label 'main_env'
		    label 'process_high'

		    input:
		    file stats_table from plot_RNA_stats_host

		    output:
		    file "*.pdf"

		    script:
		    """
		    python $workflow.projectDir/bin/plot_RNA_class_stats_each.py -i $stats_table
		    """
		}


		process plot_RNA_class_salmon_pathogen_combined {
		    publishDir "${params.outdir}/mapping_statistics/salmon/RNA_classes_pathogen", mode: 'copy'
		    storeDir "${params.outdir}/mapping_statistics/salmon/RNA_classes_pathogen"
		    tag "plot RNA class stistics combined"

		    label 'main_env'
		    label 'process_high'

		    input:
		    file stats_table from plot_RNA_stats_pathogen_combined

		    output:
		    file "RNA_class_stats_combined_pathogen.pdf"

		    script:
		    """
		    python $workflow.projectDir/bin/plot_RNA_class_stats_combined.py -i $stats_table -org pathogen
		    """
		}


		process plot_RNA_class_salmon_host_combined {
		    publishDir "${params.outdir}/mapping_statistics/salmon/RNA_classes_host", mode: 'copy'
		    storeDir "${params.outdir}/mapping_statistics/salmon/RNA_classes_host"
		    tag "plot RNA class stistics combined"

		    label 'main_env'
		    label 'process_high'

		    input:
		    file stats_table from plot_RNA_stats_host_combined

		    output:
		    file "RNA_class_stats_combined_host.pdf"

		    script:
		    """
		    python $workflow.projectDir/bin/plot_RNA_class_stats_combined.py -i $stats_table -org host
		    """
		}

}
}



/*
* STAR
*/

if(params.run_star | params.run_salmon_alignment_based_mode) {

	/*
	 * STAR - build an index
	 */


	process STARindex {
                publishDir "${params.outdir}/STAR", mode: 'copy'
		storeDir "${params.outdir}/STAR" 
		tag "STAR index"

 		label 'main_env'
         	label 'process_high'

		input:
		file(fasta) from host_pathogen_fasta_index
		file(gff) from genome_gff_star_index

		output:
		file "index/*" into star_index_transcriptome_alignment
		file "index/*" into multiqc_star_index
                file "index/*" into star_index_genome_alignment
		script:
		"""
		mkdir index
		STAR --runThreadN ${task.cpus} --runMode genomeGenerate --genomeDir index/ --genomeFastaFiles $fasta --sjdbGTFfile $gff --sjdbGTFfeatureExon quant --sjdbGTFtagExonParentTranscript Parent 
		"""
	}

}


	/*
	 * salmon - alignment_based_mode
	 */

if (params.run_salmon_alignment_based_mode){

	process ALIGNMENTstar_for_salmon {
	    tag "$sample_name"
	    publishDir "${params.outdir}/STAR_for_salmon", mode: 'copy'
	    storeDir "${params.outdir}/STAR_for_salmon" 

	    label 'main_env'
	    label 'process_medium'
	
	    input:
	    set val(sample_name),file(reads) from  trimming_results_star_salmon
	    file(gff) from gff_host_pathogen_star_salmon_alignment_gff.collect()
	    file(index) from star_index_transcriptome_alignment.collect()

	    output:
	    //file "${sample_name}/*" into multiqc_star_alignment
	    set val(sample_name), file("${sample_name}/${sample_name}Aligned.toTranscriptome.out.bam") into salmon_quantify_alignment_based_mode
	    file "${sample_name}/*"

	    script:
	outSAMunmapped = params.outSAMunmapped
	outSAMattributes = params.outSAMattributes
	quantTranscriptomeBan = params.quantTranscriptomeBan
	outFilterMultimapNmax = params.outFilterMultimapNmax
	outFilterType = params.outFilterType
	alignSJoverhangMin = params.alignSJoverhangMin
	alignSJDBoverhangMin = params.alignSJDBoverhangMin
	outFilterMismatchNmax = params.outFilterMismatchNmax
	outFilterMismatchNoverReadLmax = params.outFilterMismatchNoverReadLmax
	alignIntronMin = params.alignIntronMin
	alignIntronMax = params.alignIntronMax
	alignMatesGapMax = params.alignMatesGapMax
	    if (params.single_end){
	    """
	    mkdir $sample_name
	    STAR --runThreadN ${task.cpus} --genomeDir . --sjdbGTFfile $gff --readFilesCommand zcat --readFilesIn $reads --outSAMtype BAM Unsorted --outSAMunmapped $outSAMunmapped --outSAMattributes $outSAMattributes --outFileNamePrefix $sample_name/$sample_name --sjdbGTFfeatureExon quant --sjdbGTFtagExonParentTranscript Parent --quantMode TranscriptomeSAM --quantTranscriptomeBan $quantTranscriptomeBan --outFilterMultimapNmax $outFilterMultimapNmax --outFilterType $outFilterType --alignSJoverhangMin $alignSJoverhangMin --alignSJDBoverhangMin $alignSJDBoverhangMin --outFilterMismatchNmax $outFilterMismatchNmax --outFilterMismatchNoverReadLmax $outFilterMismatchNoverReadLmax --alignIntronMin $alignIntronMin --alignIntronMax $alignIntronMax --alignMatesGapMax $alignMatesGapMax
	    """
	    } else {
	    """
	    mkdir $sample_name
	    STAR --runThreadN ${task.cpus} --genomeDir . --sjdbGTFfile $gff --readFilesCommand zcat --readFilesIn ${reads[0]} ${reads[1]} --outSAMtype BAM Unsorted --outSAMunmapped $outSAMunmapped --outSAMattributes $outSAMattributes --outFileNamePrefix $sample_name/$sample_name --sjdbGTFfeatureExon quant --sjdbGTFtagExonParentTranscript Parent --quantMode TranscriptomeSAM -quantTranscriptomeBan $quantTranscriptomeBan --outFilterMultimapNmax $outFilterMultimapNmax --outFilterType $outFilterType --alignSJoverhangMin $alignSJoverhangMin --alignSJDBoverhangMin $alignSJDBoverhangMin --outFilterMismatchNmax $outFilterMismatchNmax --outFilterMismatchNoverReadLmax $outFilterMismatchNoverReadLmax --alignIntronMin $alignIntronMin --alignIntronMax $alignIntronMax --alignMatesGapMax $alignMatesGapMax
	    """
	    }
	}



	process salmon_quantification_alignment_based_mode {
	 //   publishDir "${params.outdir}/salmon_alignment_mode", mode: 'copy'
	    storeDir "${params.outdir}/salmon_alignment_mode"
	    tag "${sample}"

	    label 'salmon_env'
	    label 'process_high'

	    input:
	    file(transcriptome) from transcriptome_salmon_alignment_based_mode.collect()
	    set val(sample), file(bam_file) from salmon_quantify_alignment_based_mode
	    val(libtype) from libtype_salmon_alignment_mode

	    output:
	    set val(sample_name), file("${sample_name}") into split_table_alignment_based
	    file("${sample_name}") into salmon_files_to_combine_alignment_mode
//	    file("${sample_name}") into multiqc_salmon_quant
	    set val(sample_name), file("${sample_name}") into collect_processed_read_counts_alignment_based

	    script:
	    incompatPrior = params.incompatPrior
	    sample_name = sample.replaceFirst(/.fastq.gz|.fq.gz|.fastq|.fq/, "")
	    """
	    salmon quant -p ${task.cpus} -t $transcriptome -l $libtype -a $bam_file --incompatPrior $incompatPrior -o $sample_name
	    """
	}



	process split_table_salmon_each_salmon_alignment_mode {
	    publishDir "${params.outdir}/salmon_alignment_mode/${sample_name}", mode: 'copy'
	    storeDir "${params.outdir}/salmon_alignment_mode/${sample_name}"
	    tag "split_quantification_each"

	    label 'main_env'
	    label 'process_high'

	    input:
	    set val(sample_name), file ("salmon/*") from split_table_alignment_based
	    file transcriptome_pathogen from transcriptome_pathogen_to_split_q_table_salmon_alignment_based
	    file transcriptome_host from transcriptome_host_to_split_q_table_salmon_alignment_based

	    output:
	    set val(sample_name), file("host_quant.sf")
	    set val(sample_name), file("pathogen_quant.sf")

	    shell:
	    '''
	    grep ">" !{transcriptome_pathogen} | awk -F ">" '{ print $2 }' | awk 'NR==FNR{a[$0]=$0}NR>FNR{if($1==a[$1])print $0}' - salmon/*/quant.sf > pathogen_quantification_without_headers.sf
	    awk 'NR==1 {print; exit}' salmon/*/quant.sf | cat - pathogen_quantification_without_headers.sf > pathogen_quant.sf

	    grep ">" !{transcriptome_host} | awk -F ">" '{ print $2 }' | awk -F ' ' '{print $1}' | awk 'NR==FNR{a[$0]=$0}NR>FNR{if($1==a[$1])print $0}' - salmon/*/quant.sf > host_quantification_salmon_without_headers.sf
	    awk 'NR==1 {print; exit}' salmon/*/quant.sf | cat - host_quantification_salmon_without_headers.sf > host_quant.sf
	    '''
	}



	process combine_quantification_tables_salmon_alignment_mode {
	    publishDir "${params.outdir}/salmon_alignment_mode", mode: 'copy'
	    storeDir "${params.outdir}/salmon_alignment_mode"
	    tag "combine_quantification_salmon"

	    label 'main_env'
	    label 'process_high'

	    input: 
	    file input_quantification from salmon_files_to_combine_alignment_mode.collect()
	    val gene_attribute from host_atr_collect_data_salmon_alignment_mode

	    output:
	    file "combined_quant.csv" into split_table_salmon_salmon_alignment

	    script:
	    """
	    python $workflow.projectDir/bin/collect_quantification_data.py -i $input_quantification -q salmon -a $gene_attribute -org both
	    """
	}



	process split_quantification_tables_salmon_salmon_alignment_mode {
	    publishDir "${params.outdir}/salmon_alignment_mode", mode: 'copy'
	    storeDir "${params.outdir}/salmon_alignment_mode"
	    tag "split_quantification"

	    label 'main_env'
	    label 'process_high'

	    input:
	    file quant_table from split_table_salmon_salmon_alignment
	    file transcriptome_pathogen from transcriptome_pathogen_to_split_table_salmon_alignment
	    file transcriptome_host from transcriptome_host_to_split_table_salmon_alignment

	    output:
	    file 'host_quantification_salmon.csv' into host_quantification_mapping_stats_salmon_alignment_based
	    file 'pathogen_quantification_salmon.csv' into pathogen_quantification_mapping_stats_salmon_alignment_based
	    file 'host_quantification_salmon.csv' into host_quantification_RNA_stats_salmon_alignment_based
	    file 'pathogen_quantification_salmon.csv' into pathogen_quantification_RNA_stats_salmon_alignment_based
	    file 'host_quantification_salmon.csv' into quant_host_add_annotations_salmon_alignment_based
	    file 'pathogen_quantification_salmon.csv' into quant_pathogen_add_annotations_alignment_based
	    file 'host_quantification_salmon.csv' into quant_scatter_plot_host_salmon_alignment_based
	    file 'pathogen_quantification_salmon.csv' into quant_scatter_plot_pathogen_salmon_alignment_based

	    shell:
	    '''
	    grep ">" !{transcriptome_pathogen} | awk -F ">" '{ print $2 }' | awk 'NR==FNR{a[$0]=$0}NR>FNR{if($1==a[$1])print $0}' - !{quant_table}  > pathogen_quantification_without_headers.csv
	    awk 'NR==1 {print; exit}' !{quant_table} | cat - pathogen_quantification_without_headers.csv > pathogen_quantification_salmon.csv

	    grep ">" !{transcriptome_host} | awk -F ">" '{ print $2 }' | awk -F ' ' '{print $1}' | awk 'NR==FNR{a[$0]=$0}NR>FNR{if($1==a[$1])print $0}' - !{quant_table}  > host_quantification_salmon_without_headers.csv
	    awk 'NR==1 {print; exit}' !{quant_table} | cat - host_quantification_salmon_without_headers.csv > host_quantification_salmon.csv
	    '''
	}


	process combine_annotations_quant_pathogen_salmon_alignment_mode {
	    publishDir "${params.outdir}/salmon_alignment_mode", mode: 'copy'
	    storeDir "${params.outdir}/salmon_alignment_mode"
	    tag "combine_annotations_quant_pathogen"
	    label 'main_env'
	    label 'process_high'
	   
	    input: 
	    file quantification_table from quant_pathogen_add_annotations_alignment_based
	    file annotation_table from annotation_pathogen_combine_quant_salmon_alignment_based
	    val attribute from combine_annot_quant_pathogen_salmon_alignment_based

	    output:
	    file "pathogen_combined_quant_annotations.csv"

	    script:
	    """
	    $workflow.projectDir/bin/combine_quant_annotations.py -q $quantification_table -annotations $annotation_table -a $attribute -org pathogen
	    """
	}


	process combine_annotations_quant_host_salmon_alignment_mode {
	    publishDir "${params.outdir}/salmon_alignment_mode", mode: 'copy'
	    storeDir "${params.outdir}/salmon_alignment_mode"
	    tag "combine_annotations_quant_host_salmon"

	    label 'main_env'
	    label 'process_high'
	   
	    input: 
	    file quantification_table from quant_host_add_annotations_salmon_alignment_based
	    file annotation_table from annotation_host_combine_quant_salmon_alignment_based
	    val attribute from combine_annot_quant_host_salmon_alignment_based

	    output:
	    file "host_combined_quant_annotations.csv"

	    script:
	    """
	    $workflow.projectDir/bin/combine_quant_annotations.py -q $quantification_table -annotations $annotation_table -a $attribute -org host
	    """
	}




	if(params.mapping_statistics) {
		/*
		 * salmon - alignment-based 'quantification_stats'
		 */

		process scatter_plot_pathogen_salmon_alignment_based {
		    publishDir "${params.outdir}/mapping_statistics/salmon_alignment_based/scatter_plots", mode: 'copy'
		    storeDir "${params.outdir}/mapping_statistics/salmon_alignment_based/scatter_plots"
		    tag "scatter_plot salmon"

		    label 'main_env'
		    label 'process_high'

		    input:
		    file quant_table from quant_scatter_plot_pathogen_salmon_alignment_based
		    val attribute from atr_scatter_plot_pathogen_alignment

		    output:
		    file ('*.pdf')

		    script:
		    """
		    python $workflow.projectDir/bin/scatter_plots.py -q $quant_table -a $attribute -org pathogen
		    """
		}


	process scatter_plot_host_salmon_alignment_based {
	    publishDir "${params.outdir}/mapping_statistics/salmon_alignment_based/scatter_plots", mode: 'copy'
	    storeDir "${params.outdir}/mapping_statistics/salmon_alignment_based/scatter_plots"
		    tag "scatter_plot salmon"

		    label 'main_env'
		    label 'process_high'

		    input:
		    file quant_table from quant_scatter_plot_host_salmon_alignment_based
		    val attribute from atr_scatter_plot_host_alignment

		    output:
		    file ('*.pdf') 

		    script:
		    """
		    python $workflow.projectDir/bin/scatter_plots.py -q $quant_table -a $attribute -org host 
		    """
		}


	process extract_processed_reads_salmon_alignment_based {
		    publishDir "${params.outdir}/mapping_statistics/salmon_alignment_based", mode: 'copy'
		    storeDir "${params.outdir}/mapping_statistics/salmon_alignment_based"
		    tag "extract_processed_reads"

		    label 'main_env'
		    label 'process_high'
		   
		    input: 
		    set val(sample_name), file ("salmon/*") from collect_processed_read_counts_alignment_based

		    output:
		    file "${sample_name}.txt" into collect_results_alignment_based

		    script:
		    """
		    $workflow.projectDir/bin/extract_processed_reads_salmon.sh salmon/*/aux_info/meta_info.json $sample_name
		    """
		}


	process collect_processed_reads_salmon_alignment_based {
		    publishDir "${params.outdir}/mapping_statistics/salmon_alignment_based", mode: 'copy'
		    storeDir "${params.outdir}/mapping_statistics/salmon_alignment_based"
		    tag "collect_processed_reads"

		    label 'main_env'
		    label 'process_high' 
		    
		    input: 
		    file process_reads from collect_results_alignment_based.collect()

		    output:
		    file "processed_reads_salmon.csv" into mapping_stats_total_reads_alignment
		    script:
		    """
		    cat $process_reads > processed_reads_salmon.csv
		    """
		}



		process salmon_quantification_stats_salmon_alignment_based {
		    publishDir "${params.outdir}/mapping_statistics/salmon_alignment_based", mode: 'copy'
		    storeDir "${params.outdir}/mapping_statistics/salmon_alignment_based"
		    tag "quantification_statistics salmon"

		    label 'main_env'
		    label 'process_high'

		    input:
		    file quant_table_host from host_quantification_mapping_stats_salmon_alignment_based
		    file quant_table_pathogen from pathogen_quantification_mapping_stats_salmon_alignment_based
		    val attribute from attribute_quant_stats_salmon_alignment
		    file total_processed_reads from mapping_stats_total_reads_alignment
		    file total_raw_reads from collect_total_reads_raw_salmon_alignment

		    output:
		    file ('salmon_host_pathogen_total_reads.csv') into salmon_mapped_stats_to_plot_alignment

		    script:
		    """
		    python $workflow.projectDir/bin/quantification_stats.py -q_p $quant_table_pathogen -q_h $quant_table_host -total_processed $total_processed_reads -total_raw $total_raw_reads -a $attribute -t salmon -o salmon_host_pathogen_total_reads.csv
		    """
		}


		process plot_salmon_mapping_stats_host_pathogen_salmon_alignment_based {
		    tag "$name2"
		    publishDir "${params.outdir}/mapping_statistics/salmon_alignment_based", mode: 'copy'
		    storeDir "${params.outdir}/mapping_statistics/salmon_alignment_based"

		    label 'main_env'
		    label 'process_high'

		    input:
		    file(stats) from salmon_mapped_stats_to_plot_alignment

		    output:
		    file "mapping_stats_*"

		    script:
		    """
		    python $workflow.projectDir/bin/plot_mapping_statistics.py -i $stats
		    """
		}


		process RNA_class_statistics_salmon_pathogen_alignment_based {
		    publishDir "${params.outdir}/mapping_statistics/salmon_alignment_based/RNA_classes_pathogen", mode: 'copy'
		    storeDir "${params.outdir}/mapping_statistics/salmon_alignment_based/RNA_classes_pathogen"
		    tag "RNA class stistics"

		    label 'main_env'
		    label 'process_high'

		    input:
		    file quant_table from pathogen_quantification_RNA_stats_salmon_alignment_based
		    val attribute from host_annotations_RNA_class_stats_pathogen_alignment
		    file gene_annotations from pathogen_annotations_RNA_class_stats_salmon_alignment

		    output:
		    file "pathogen_RNA_classes_percentage.csv" into plot_RNA_stats_pathogen_alignment
		    file "pathogen_RNA_classes_percentage.csv" into plot_RNA_stats_pathogen_combined_alignment
		    file "pathogen_RNA_classes_sum_counts.csv"

		    script:
		    """
		    python $workflow.projectDir/bin/RNA_class_content.py -q $quant_table -a $attribute -annotations $gene_annotations -q_tool salmon -org pathogen
		    """
		}


		process RNA_class_statistics_salmon_host_alignment_based {
		    publishDir "${params.outdir}/mapping_statistics/salmon_alignment_based/RNA_classes_host", mode: 'copy'
		    storeDir "${params.outdir}/mapping_statistics/salmon_alignment_based/RNA_classes_host"
		    tag "RNA class stistics"

		    label 'main_env'
		    label 'process_high'

		    input:
		    file quant_table from host_quantification_RNA_stats_salmon_alignment_based
		    val attribute from attribute_host_RNA_class_stats_alignment
		    file gene_annotations from host_annotations_RNA_class_stats_salmon_alignment
		    file rna_classes_to_replace from RNA_classes_to_replace_alignment

		    output:
		    file "host_RNA_classes_percentage.csv" into plot_RNA_stats_host_alignment
		    file "host_RNA_classes_percentage.csv" into plot_RNA_stats_host_combined_alignment
		    file "host_RNA_classes_sum_counts.csv"
		    file "host_gene_types_groups_*"

		    script:
		    """
		    python $workflow.projectDir/bin/RNA_class_content.py -q $quant_table -a $attribute -annotations $gene_annotations -rna $rna_classes_to_replace -q_tool salmon -org host
		    """
		}



		process plot_RNA_class_salmon_pathogen_each_alignment_based{
		    publishDir "${params.outdir}/mapping_statistics/salmon_alignment_based/RNA_classes_pathogen", mode: 'copy'
		    storeDir "${params.outdir}/mapping_statistics/salmon_alignment_based/RNA_classes_pathogen"
		    tag "plot RNA class stistics each"

		    label 'main_env'
		    label 'process_high'

		    input:
		    file stats_table from plot_RNA_stats_pathogen_alignment

		    output:
		    file "*.pdf"

		    script:
		    """
		    python $workflow.projectDir/bin/plot_RNA_class_stats_each.py -i $stats_table
		    """
		}

		process plot_RNA_class_salmon_host_each_alignment_based {
		    publishDir "${params.outdir}/mapping_statistics/salmon_alignment_based/RNA_classes_host", mode: 'copy'
		    storeDir "${params.outdir}/mapping_statistics/salmon_alignment_based/RNA_classes_host"
		    tag "plot RNA class stistics each"

		    label 'main_env'
		    label 'process_high'

		    input:
		    file stats_table from plot_RNA_stats_host_alignment

		    output:
		    file "*.pdf"

		    script:
		    """
		    python $workflow.projectDir/bin/plot_RNA_class_stats_each.py -i $stats_table
		    """
		}


		process plot_RNA_class_salmon_pathogen_combined_alignment_based {
		    publishDir "${params.outdir}/mapping_statistics/salmon_alignment_based/RNA_classes_pathogen", mode: 'copy'
		    storeDir "${params.outdir}/mapping_statistics/salmon_alignment_based/RNA_classes_pathogen"
		    tag "plot RNA class stistics combined"

		    label 'main_env'
		    label 'process_high'

		    input:
		    file stats_table from plot_RNA_stats_pathogen_combined_alignment

		    output:
		    file "RNA_class_stats_combined_pathogen.pdf"

		    script:
		    """
		    python $workflow.projectDir/bin/plot_RNA_class_stats_combined.py -i $stats_table -org pathogen
		    """
		}


		process plot_RNA_class_salmon_host_combined_alignment_based {
		    publishDir "${params.outdir}/mapping_statistics/salmon_alignment_based/RNA_classes_host", mode: 'copy'
		    storeDir "${params.outdir}/mapping_statistics/salmon_alignment_based/RNA_classes_host"
		    tag "plot RNA class stistics combined"

		    label 'main_env'
		    label 'process_high'

		    input:
		    file stats_table from plot_RNA_stats_host_combined_alignment

		    output:
		    file "RNA_class_stats_combined_host.pdf"

		    script:
		    """
		    python $workflow.projectDir/bin/plot_RNA_class_stats_combined.py -i $stats_table -org host
		    """
		}

}

}


if(params.run_star) {

	/*
	 * STAR - align the reads
	 */

	process ALIGNMENTstar {
	    tag "${sample_name}"
            publishDir "${params.outdir}/STAR", mode: 'copy'
	    storeDir "${params.outdir}/STAR" 

            label 'main_env'
            label 'process_high'
	
	    input:
	    set val(sample_name),file(reads) from  trimming_results_star_htseq
	    file(gff) from gff_host_pathogen_star_htseq_alignment_gff.collect()
	    file(index) from star_index_genome_alignment.collect()

	    output:
//	    set val(sample_name), file("${sample_name}/${sample_name}Aligned*.out.bam") into star_aligned_h_p
//	    file("${sample_name}/${sample_name}Aligned*.out.bam") into star_aligned_h_p2
	    set val(sample_name), file("${sample_name}/${sample_name}Aligned*.out.bam") into star_aligned_u_m
//	    file("${sample_name}/${sample_name}Aligned*.out.bam") into star_aligned_m_m
	    set val(sample_name), file("${sample_name}/${sample_name}Aligned*.out.bam") into alignment_unique_mapping_stats
	    set val(sample_name), file("${sample_name}/${sample_name}Aligned*.out.bam") into alignment_crossmapped_extract
	    set val(sample_name), file("${sample_name}/${sample_name}Aligned*.out.bam") into alignment_crossmapped_find
//	    file "${sample_name}/*" into multiqc_star_alignment

	    script:
	outSAMunmapped = params.outSAMunmapped
	outSAMattributes = params.outSAMattributes
	outFilterMultimapNmax = params.outFilterMultimapNmax
	outFilterType = params.outFilterType
	alignSJoverhangMin = params.alignSJoverhangMin
	alignSJDBoverhangMin = params.alignSJDBoverhangMin
	outFilterMismatchNmax = params.outFilterMismatchNmax
	outFilterMismatchNoverReadLmax = params.outFilterMismatchNoverReadLmax
	alignIntronMin = params.alignIntronMin
	alignIntronMax = params.alignIntronMax
	alignMatesGapMax = params.alignMatesGapMax
	outWigType = params.outWigType
	outWigStrand = params.outWigStrand
	    if (params.single_end){
	    """
	    mkdir $sample_name
	    STAR --runThreadN ${task.cpus} --genomeDir . --sjdbGTFfile $gff --readFilesCommand zcat --readFilesIn $reads --outSAMtype BAM SortedByCoordinate --outSAMunmapped $outSAMunmapped --outSAMattributes $outSAMattributes --outWigType $outWigType --outWigStrand $outWigStrand --outFileNamePrefix $sample_name/$sample_name --sjdbGTFfeatureExon quant --sjdbGTFtagExonParentTranscript Parent --outFilterMultimapNmax $outFilterMultimapNmax --outFilterType $outFilterType --alignSJoverhangMin $alignSJoverhangMin --alignSJDBoverhangMin $alignSJDBoverhangMin --outFilterMismatchNmax $outFilterMismatchNmax --outFilterMismatchNoverReadLmax $outFilterMismatchNoverReadLmax --alignIntronMin $alignIntronMin --alignIntronMax $alignIntronMax --alignMatesGapMax $alignMatesGapMax
	    """
	    } else {
	    """
	    mkdir $sample_name
	    STAR --runThreadN ${task.cpus} --genomeDir . --sjdbGTFfile $gff --readFilesCommand zcat --readFilesIn ${reads[0]} ${reads[1]} --outSAMtype BAM SortedByCoordinate --outSAMunmapped $outSAMunmapped --outSAMattributes $outSAMattributes --outWigType $outWigType --outWigStrand outWigStrand --outFileNamePrefix $sample_name/$sample_name --sjdbGTFfeatureExon quant --sjdbGTFtagExonParentTranscript Parent --outFilterMultimapNmax $outFilterMultimapNmax --outFilterType $outFilterType --alignSJoverhangMin $alignSJoverhangMin --alignSJDBoverhangMin $alignSJDBoverhangMin --outFilterMismatchNmax $outFilterMismatchNmax --outFilterMismatchNoverReadLmax $outFilterMismatchNoverReadLmax --alignIntronMin $alignIntronMin --alignIntronMax $alignIntronMax --alignMatesGapMax $alignMatesGapMax
	    """
	    }
	}



	if(params.run_htseq_multi_mapped | params.mapping_statistics) {
		/*
		 * find_cross_mapped_reads 
		*/

		process find_crossmapped_reads {
		    tag "$sample_name"
		    publishDir "${params.outdir}/STAR/multimapped_reads", mode: 'copy'
		    storeDir "${params.outdir}/STAR/multimapped_reads"
		    
                    label 'main_env'
                    label 'process_high'

		    input:
		    set val(sample_name), file(alignment) from alignment_crossmapped_find
		    file(host_reference) from reference_host_names_crossmapped_find.collect()
		    file(pathogen_reference) from reference_pathogen_crossmapped_find.collect()

		    output:
		    file "${cross_mapped_reads}" into remove_crossmapped_reads
		    file "${cross_mapped_reads}" into count_crossmapped_reads

		    shell:
		    cross_mapped_reads = sample_name + "_cross_mapped_reads.txt"
		    '''
		    samtools view -F 4 -h !{alignment} | fgrep -vw NH:i:1 | !{workflow.projectDir}/bin/extract_crossmapped_reads.py -h_ref !{host_reference} -p_ref !{pathogen_reference} -o !{cross_mapped_reads}
		    '''
		}


		/*
		 * remove_cross_mapped_reads
		 */

		process remove_crossmapped_reads {
		    tag "$sample_name"
		    publishDir "${params.outdir}/STAR/multimapped_reads", mode: 'copy'
		    storeDir "${params.outdir}/STAR/multimapped_reads"

                    label 'main_env'
                    label 'process_high'

		    input:
		    set val(sample_name), file(alignment) from alignment_crossmapped_extract
		    file(cross_mapped_reads) from remove_crossmapped_reads

		    output:
		    file "${bam_file_without_crossmapped}" into alignment_multi_mapping_stats
//		    file "${bam_file_without_crossmapped}" into without_crossmapped_m_m

		    shell:
		    bam_file_without_crossmapped = sample_name + "_no_crossmapped.bam"
		    '''
		    samtools view -@ !{task.cpus} -h !{alignment} | fgrep -vf !{cross_mapped_reads} | samtools view -@ !{task.cpus} -bS -o !{bam_file_without_crossmapped} -
		    '''
		}
	}


	if(params.mapping_statistics) {

		/*
		 * uniquly mapped reads - statistics
		 */


		process unique_mapping_stats_STAR {
		    tag "$name"
		    publishDir "${params.outdir}/mapping_statistics/STAR/uniquely_mapped", mode: 'copy'
		    storeDir "${params.outdir}/mapping_statistics/STAR/uniquely_mapped"

		    label 'main_env'
		    label 'process_high'

		    input:
		    set val(sample_name), file(alignment) from alignment_unique_mapping_stats
		    file(host_reference_names) from reference_host_names_uniquelymapped.collect()
		    file(pathogen_reference_names) from reference_pathogen_names_uniquelymapped.collect()

		    output:
		    file "${out_file_name}" into unique_stats_to_plot
		   
		    shell: 
		    name = alignment[0].toString()
		    out_file_name = name.replaceFirst(/Aligned.sortedByCoord.out.bam/, "uniquly_mapped_reads_statistics.txt")
		    '''
		    samtools view -F 4 -h !{alignment} | grep -f !{host_reference_names} | fgrep -w NH:i:1 | echo "!{sample_name} host `wc -l`" > host_uniquly_mapped_reads_sum.txt
		    samtools view -F 4 -h !{alignment} | grep -f !{pathogen_reference_names} | fgrep -w NH:i:1 | echo "!{sample_name} pathogen `wc -l`" > pathogen_uniquly_mapped_reads_sum.txt
		    cat host_uniquly_mapped_reads_sum.txt pathogen_uniquly_mapped_reads_sum.txt >> !{out_file_name}
		    '''
		}

		/*
		 * count_cross_mapped_reads 
		*/

		process count_crossmapped_reads {
		    tag "$name2"
		    publishDir "${params.outdir}/mapping_statistics/STAR/multi_mapped", mode: 'copy'
		    storeDir "${params.outdir}/mapping_statistics/STAR/multi_mapped"

		    label 'process_high'

		    input:
		    file(cross_mapped_reads) from count_crossmapped_reads.collect()

		    output:
                    file "cross_mapped_reads_sum.txt"
                    file "cross_mapped_reads_sum2.txt"
		    
		    shell:
		    '''
		    wc -l !{cross_mapped_reads} >> cross_mapped_reads_sum.txt
		    head -n -1 cross_mapped_reads_sum.txt > cross_mapped_reads_sum2.txt
		    '''
		}


		/*
		 * multi mapped reads - statistics (uniquely_mapped + multi_mapped - cross_mapped reads )
		 */

		process multi_mapping_stats {
		    tag "$name2"
		    storeDir "${params.outdir}/mapping_statistics/STAR/multi_mapped"

                    label 'main_env'
		    label 'process_high'

		    input:
		    file(alignment) from alignment_multi_mapping_stats
		    file(host_reference_names) from reference_host_names_multimapped.collect()
		    file(pathogen_reference_names) from reference_pathogen_names_multimapped.collect()

		    output:
		    file "${out_file_name}" into multi_mapped_stats_to_plot

		    shell: 
		    name = alignment[0].toString()
		    name2 = name.replaceFirst(/_no_crossmapped.bam/, "") 
		    out_file_name = name.replaceFirst(/no_crossmapped.bam/, "multi_mapped_reads_statistics.txt")
		    '''
		    samtools view -F 4 -h !{alignment} | grep -f !{host_reference_names} | fgrep -w HI:i:1| echo "!{name2} host `wc -l`" > host_multi_mapped_reads_sum.txt
		    samtools view -F 4 -h !{alignment} | grep -f !{pathogen_reference_names} | fgrep -w HI:i:1 | echo "!{name2} pathogen `wc -l`" > pathogen_multi_mapped_reads_sum.txt
		    cat host_multi_mapped_reads_sum.txt pathogen_multi_mapped_reads_sum.txt >> !{out_file_name}
		    '''
		}


}
}



/*
* htseq
*/


if(params.run_htseq_uniquely_mapped){


	/*
	 * HTseq - quantification -unique mapping
	 */

	process HTseq_unique_mapping {
	    publishDir "${params.outdir}/HTSeq/uniquely_mapped", mode: 'copy'
	    storeDir "${params.outdir}/HTSeq/uniquely_mapped"
	    tag "$sample_name"

            label 'main_env'
            label 'process_high'

	    input:
	    file(gff) from quantification_gff_u_m.collect()
	    set val(sample_name), file(st) from star_aligned_u_m
	    val(host_attribute) from host_gff_attribute_htseq

	    output:
//	    file ("$name_file2") into htseq_files_u_m
	    file ("$name_file2") into htseq_files_to_combine
//	    file ("$name_file2") into multiqc_htseq_unique

	    script:
	    name_file2 = sample_name + "_count_u_m"
	    host_attr = host_attribute
	    """
	    htseq-count -t quant -f bam -r pos $st $gff -i $host_attr > $name_file2
	    sed -i '1{h;s/.*/'"$sample_name"'/;G}' "$name_file2"
	    """
	}



	/*
	 * htseq - combine quantification results
	 */

	htseq_files_to_combine
	    .collect()
	    .set { htseq_result_quantification }

	process combine_quantification_tables_htseq_uniquely_mapped {
		    publishDir "${params.outdir}/HTSeq/uniquely_mapped", mode: 'copy'
		    storeDir "${params.outdir}/HTSeq/uniquely_mapped"
		    tag "combine_quantification_htseq_uniquely_mapped"

                    label 'process_high'

		    input: 
		    file input_quantification from htseq_result_quantification
		    val(host_attribute) from host_gff_attribute_htseq_combine

		    output:
		    file "quantification_results_uniquely_mapped.csv" into split_table_htseq_host
		    file "quantification_results_uniquely_mapped.csv" into split_table_htseq_pathogen
		    file "alignment_stats_uniquely_mapped.csv"

		    script:
		    """
		    python $workflow.projectDir/bin/collect_quantification_data.py -i $input_quantification -q htseq -a $host_attribute -p uniquely_mapped
		    """
		}
 

	/*
	 * htseq - split quantification tables into host and pathogen results
	 */


	process split_quantification_tables_htseq_uniquely_mapped_host {
		    publishDir "${params.outdir}/HTSeq/uniquely_mapped", mode: 'copy'
		    storeDir "${params.outdir}/HTSeq/uniquely_mapped"
		    tag "split_quantification_htseq_uniquely_mapped_host"

                    label 'process_high'

		    input:
		    file quant_table from split_table_htseq_host
		    file host_annotations from annotation_host_split_quant_htseq

		    output:
		    file 'host_quantification_uniquely_mapped_htseq.csv' into host_quantification_stats_htseq
		    file 'host_quantification_uniquely_mapped_htseq.csv' into host_quantification_stats_htseq_total
		    file 'host_quantification_uniquely_mapped_htseq.csv' into host_htseq_quantification_RNA_stats

		    shell:
		    '''
		    awk -F"\t" '!(NR<=1) {print $1}' !{host_annotations}| awk 'NR==FNR{a[$0]=$0}NR>FNR{if($1==a[$1])print $0}' - !{quant_table} > host_quantification_uniquely_mapped_htseq_without_headers.csv
		    awk 'NR==1 {print; exit}' !{quant_table} | cat - host_quantification_uniquely_mapped_htseq_without_headers.csv > host_quantification_uniquely_mapped_htseq.csv
		    '''
		}


	process split_quantification_tables_htseq_uniquely_mapped_pathogen {
		    publishDir "${params.outdir}/HTSeq/uniquely_mapped", mode: 'copy'
		    storeDir "${params.outdir}/HTSeq/uniquely_mapped"
		    tag "split_quantification_htseq_uniquely_mapped_pathogen"

                    label 'main_env'

		    input:
		    file quant_table from split_table_htseq_pathogen
		    file pathogen_annotations from annotation_pathogen_split_quant_htseq

		    output:
		    file 'pathogen_quantification_uniquely_mapped_htseq.csv' into pathogen_quantification_stats_htseq
		    file 'pathogen_quantification_uniquely_mapped_htseq.csv' into pathogen_quantification_stats_htseq_total
		    file 'pathogen_quantification_uniquely_mapped_htseq.csv' into pathogen_htseq_quantification_RNA_stats
		    file 'pathogen_quantification_uniquely_mapped_htseq.csv' into quant_pathogen_add_annotations_htseq_u_m

		    shell:
		    '''
		    awk -F"\t" '!(NR<=1) {print $1}' !{pathogen_annotations}| awk 'NR==FNR{a[$0]=$0}NR>FNR{if($1==a[$1])print $0}' - !{quant_table} > pathogen_quantification_uniquely_mapped_htseq_without_headers.csv
		    awk 'NR==1 {print; exit}' !{quant_table} | cat - pathogen_quantification_uniquely_mapped_htseq_without_headers.csv > pathogen_quantification_uniquely_mapped_htseq.csv
		    '''
		}


		process combine_annotations_quant_pathogen_uniquely_mapped_host {
		    publishDir "${params.outdir}/HTSeq/uniquely_mapped", mode: 'copy'
		    storeDir "${params.outdir}/HTSeq/uniquely_mapped"
		    tag "combine_annotations_quant_pathogen"
		    label 'main_env'
		    label 'process_high'
		   
		    input: 
		    file quantification_table from quant_pathogen_add_annotations_htseq_u_m
		    file annotation_table from annotation_pathogen_combine_quant_htseq_u_m
		    val attribute from combine_annot_quant_pathogen_host_gff_attribute

		    output:
		    file "pathogen_combined_quant_annotations.csv"

		    script:
		    """
		    $workflow.projectDir/bin/combine_quant_annotations.py -q $quantification_table -annotations $annotation_table -a $attribute -org pathogen
		    """
		}


	process combine_annotations_quant_host_uniquely_mapped_host {
	    publishDir "${params.outdir}/HTSeq/uniquely_mapped", mode: 'copy'
	    storeDir "${params.outdir}/HTSeq/uniquely_mapped"
	    tag "combine_annotations_quant_host_salmon"

	    label 'main_env'
	    label 'process_high'
	   
	    input: 
	    file quantification_table from quant_host_add_annotations_salmon_alignment_based
	    file annotation_table from annotation_host_combine_quant_salmon_alignment_based
	    val attribute from combine_annot_quant_host_salmon_alignment_based

	    output:
	    file "host_combined_quant_annotations.csv"

	    script:
	    """
	    $workflow.projectDir/bin/combine_quant_annotations.py -q $quantification_table -annotations $annotation_table -a $attribute -org host
	    """
	}


	if(params.mapping_statistics) {
		process htseq_quantification_stats_uniquely_mapped {
		    storeDir "${params.outdir}/mapping_statistics/HTSeq/uniquely_mapped"
		    publishDir "${params.outdir}/mapping_statistics/HTSeq/uniquely_mapped", mode: 'copy'
		    tag "quantification_statistics htseq"

		    label 'main_env'

		    input:
		    file quant_table_host from host_quantification_stats_htseq_total
		    file quant_table_pathogen from pathogen_quantification_stats_htseq_total
		    file total_raw_reads from collect_total_reads_raw_htseq_uniquely_mapped
		    val attribute from host_gff_attribute_mapping_stats_htseq
		    file total_processed_reads from mapping_stats_total_processed_reads_alignment

		    output:
		    file ('htseq_uniquely_mapped_host_pathogen_total_reads.csv') into htseq_mapped_stats_to_plot

		    script:
		    """
		    python $workflow.projectDir/bin/quantification_stats.py -q_p $quant_table_pathogen -q_h $quant_table_host -a $attribute -total_raw $total_raw_reads -total_processed $total_processed_reads -t htseq -o htseq_uniquely_mapped_host_pathogen_total_reads.csv
		    """
		}


		process plot_mapping_stats_host_pathogen_htseq_uniquely_mapped{
		    tag "$name2"
		    publishDir "${params.outdir}/mapping_statistics/HTSeq/uniquely_mapped", mode: 'copy'
		    storeDir "${params.outdir}/mapping_statistics/HTSeq/uniquely_mapped"

		    label 'main_env'
		    label 'process_high'

		    input:
		    file(stats) from htseq_mapped_stats_to_plot

		    output:
		    file "mapping_stats_*"

		    script:
		    """
		    python $workflow.projectDir/bin/plot_mapping_statistics.py -i $stats
		    """
		}


		process RNA_class_statistics_htseq_uniquely_mapped_pathogen {
		    storeDir "${params.outdir}/mapping_statistics/HTSeq/uniquely_mapped/RNA_classes_pathogen/"
		    publishDir "${params.outdir}/mapping_statistics/HTSeq/uniquely_mapped/RNA_classes_pathogen", mode: 'copy'
		    tag "RNA class stistics htseq pathogen "

		    label 'main_env'
		    label 'process_high'

		    input:
		    file quant_table from pathogen_htseq_quantification_RNA_stats
		    val attribute from host_gff_attribute_RNA_class_pathogen_htseq
		    file gene_annotations from pathogen_annotations_RNA_class_stats_htseq


		    output:
		    file "pathogen_RNA_classes_percentage_uniquely_mapped.csv" into plot_RNA_stats_pathogen_htseq_u_m
		    file "pathogen_RNA_classes_percentage_uniquely_mapped.csv" into plot_RNA_stats_pathogen_combined_htseq_u_m
		    file "pathogen_RNA_classes_sum_counts_uniquely_mapped.csv"


		    script:
		    """
		    python $workflow.projectDir/bin/RNA_class_content.py -q $quant_table -a $attribute -annotations $gene_annotations -q_tool htseq -org pathogen -p uniquely_mapped
		    """
		}


		process RNA_class_statistics_htseq_uniquely_mapped_host {
		    publishDir "${params.outdir}/mapping_statistics/HTSeq/uniquely_mapped/RNA_classes_host", mode: 'copy'
		    storeDir "${params.outdir}/mapping_statistics/HTSeq/uniquely_mapped/RNA_classes_host"
		    tag "RNA class stistics htseq host"

		    label 'main_env'
		    label 'process_high'

		    input:
		    file quant_table from host_htseq_quantification_RNA_stats
		    val attribute from host_gff_attribute_RNA_class_host_htseq
		    file gene_annotations from host_annotations_RNA_class_stats_htseq
		    file rna_classes_to_replace from RNA_classes_to_replace_htseq_uniquely_mapped

		    output:
		    file "host_RNA_classes_percentage_uniquely_mapped.csv" into plot_RNA_stats_host_htseq_u_m
		    file "host_RNA_classes_percentage_uniquely_mapped.csv" into plot_RNA_stats_host_combined_htseq_u_m
		    file "host_RNA_classes_sum_counts_uniquely_mapped.csv"
		    file "host_gene_types_groups_*"

		    script:
		    """
		    python $workflow.projectDir/bin/RNA_class_content.py -q $quant_table -a $attribute -annotations $gene_annotations -rna $rna_classes_to_replace -q_tool htseq -org host -p uniquely_mapped
		    """
		}



		process plot_RNA_class_htseq_uniquely_mapped_pathogen_each{
		    publishDir "${params.outdir}/mapping_statistics/HTSeq/uniquely_mapped/RNA_classes_pathogen", mode: 'copy'
		    storeDir "${params.outdir}/mapping_statistics/HTSeq/uniquely_mapped/RNA_classes_pathogen"
		    tag "plot RNA class stistics each"

		    label 'main_env'
		    label 'process_high'

		    input:
		    file stats_table from plot_RNA_stats_pathogen_htseq_u_m

		    output:
		    file "*.pdf"

		    script:
		    """
		    python $workflow.projectDir/bin/plot_RNA_class_stats_each.py -i $stats_table
		    """
		}

		process plot_RNA_class_htseq_uniquely_mapped_host_each {
		    publishDir "${params.outdir}/mapping_statistics/HTSeq/uniquely_mapped/RNA_classes_host", mode: 'copy'
		    storeDir "${params.outdir}/mapping_statistics/HTSeq/uniquely_mapped/RNA_classes_host"
		    tag "plot RNA class stistics each"

		    label 'main_env'
		    label 'process_high'

		    input:
		    file stats_table from plot_RNA_stats_host_htseq_u_m

		    output:
		    file "*.pdf"

		    script:
		    """
		    python $workflow.projectDir/bin/plot_RNA_class_stats_each.py -i $stats_table
		    """
		}


		process plot_RNA_class_htseq_uniquely_mapped_pathogen_combined {
		    publishDir "${params.outdir}/mapping_statistics/HTSeq/uniquely_mapped/RNA_classes_pathogen", mode: 'copy'
		    storeDir "${params.outdir}/mapping_statistics/HTSeq/uniquely_mapped/RNA_classes_pathogen"
		    tag "plot RNA class stistics combined"

		    label 'main_env'
		    label 'process_high'

		    input:
		    file stats_table from plot_RNA_stats_pathogen_combined_htseq_u_m

		    output:
		    file "RNA_class_stats_combined_pathogen.pdf"

		    script:
		    """
		    python $workflow.projectDir/bin/plot_RNA_class_stats_combined.py -i $stats_table -org pathogen
		    """
		}


		process plot_RNA_class_htseq_uniquely_host_combined {
		    publishDir "${params.outdir}/mapping_statistics/HTSeq/uniquely_mapped/RNA_classes_host", mode: 'copy'
		    storeDir "${params.outdir}/mapping_statistics/HTSeq/uniquely_mapped/RNA_classes_host"
		    tag "plot RNA class stistics combined"

		    label 'main_env'
		    label 'process_high'

		    input:
		    file stats_table from plot_RNA_stats_host_combined_htseq_u_m

		    output:
		    file "RNA_class_stats_combined_host.pdf"

		    script:
		    """
		    python $workflow.projectDir/bin/plot_RNA_class_stats_combined.py -i $stats_table -org host
		    """
		}




}



}


























/*
 * STEP 2 - MultiQC
 */
/*
process multiqc {
    publishDir "${params.outdir}/MultiQC", mode: 'copy'
    storeDir "${params.outdir}/MultiQC"

    input:
    file (multiqc_config) from ch_multiqc_config
    file (mqc_custom_config) from ch_multiqc_custom_config.collect().ifEmpty([])
    // TODO nf-core: Add in log files from your new processes for MultiQC to find!
    file ('fastqc/*') from ch_fastqc_results.collect().ifEmpty([])
    file ('software_versions/*') from ch_software_versions_yaml.collect()
    file workflow_summary from ch_workflow_summary.collectFile(name: "workflow_summary_mqc.yaml")

    output:
    file "*multiqc_report.html" into ch_multiqc_report
    file "*_data"
    file "multiqc_plots"

    script:
    rtitle = custom_runName ? "--title \"$custom_runName\"" : ''
    rfilename = custom_runName ? "--filename " + custom_runName.replaceAll('\\W','_').replaceAll('_+','_') + "_multiqc_report" : ''
    custom_config_file = params.multiqc_config ? "--config $mqc_custom_config" : ''
    // TODO nf-core: Specify which MultiQC modules to use with -m for a faster run time
    """
    multiqc -f $rtitle $rfilename $custom_config_file .
    """
}
*/

/*
 * STEP 3 - Output Description HTML
 */

/*
process output_documentation {
    publishDir "${params.outdir}/pipeline_info", mode: 'copy'

    input:
    file output_docs from ch_output_docs

    output:
    file "results_description.html"

    script:
    """
    markdown_to_html.py $output_docs -o results_description.html
    """
}

*/



/*
 * Completion e-mail notification
 */
workflow.onComplete {

    // Set up the e-mail variables
    def subject = "[nf-core/dualrnaseq] Successful: $workflow.runName"
    if (!workflow.success) {
        subject = "[nf-core/dualrnaseq] FAILED: $workflow.runName"
    }
    def email_fields = [:]
    email_fields['version'] = workflow.manifest.version
    email_fields['runName'] = custom_runName ?: workflow.runName
    email_fields['success'] = workflow.success
    email_fields['dateComplete'] = workflow.complete
    email_fields['duration'] = workflow.duration
    email_fields['exitStatus'] = workflow.exitStatus
    email_fields['errorMessage'] = (workflow.errorMessage ?: 'None')
    email_fields['errorReport'] = (workflow.errorReport ?: 'None')
    email_fields['commandLine'] = workflow.commandLine
    email_fields['projectDir'] = workflow.projectDir
    email_fields['summary'] = summary
    email_fields['summary']['Date Started'] = workflow.start
    email_fields['summary']['Date Completed'] = workflow.complete
    email_fields['summary']['Pipeline script file path'] = workflow.scriptFile
    email_fields['summary']['Pipeline script hash ID'] = workflow.scriptId
    if (workflow.repository) email_fields['summary']['Pipeline repository Git URL'] = workflow.repository
    if (workflow.commitId) email_fields['summary']['Pipeline repository Git Commit'] = workflow.commitId
    if (workflow.revision) email_fields['summary']['Pipeline Git branch/tag'] = workflow.revision
    email_fields['summary']['Nextflow Version'] = workflow.nextflow.version
    email_fields['summary']['Nextflow Build'] = workflow.nextflow.build
    email_fields['summary']['Nextflow Compile Timestamp'] = workflow.nextflow.timestamp

    // TODO nf-core: If not using MultiQC, strip out this code (including params.max_multiqc_email_size)
    // On success try attach the multiqc report
    def mqc_report = null
    try {
        if (workflow.success) {
            mqc_report = ch_multiqc_report.getVal()
            if (mqc_report.getClass() == ArrayList) {
                log.warn "[nf-core/dualrnaseq] Found multiple reports from process 'multiqc', will use only one"
                mqc_report = mqc_report[0]
            }
        }
    } catch (all) {
        log.warn "[nf-core/dualrnaseq] Could not attach MultiQC report to summary email"
    }

    // Check if we are only sending emails on failure
    email_address = params.email
    if (!params.email && params.email_on_fail && !workflow.success) {
        email_address = params.email_on_fail
    }

    // Render the TXT template
    def engine = new groovy.text.GStringTemplateEngine()
    def tf = new File("$baseDir/assets/email_template.txt")
    def txt_template = engine.createTemplate(tf).make(email_fields)
    def email_txt = txt_template.toString()

    // Render the HTML template
    def hf = new File("$baseDir/assets/email_template.html")
    def html_template = engine.createTemplate(hf).make(email_fields)
    def email_html = html_template.toString()

    // Render the sendmail template
    def smail_fields = [ email: email_address, subject: subject, email_txt: email_txt, email_html: email_html, baseDir: "$baseDir", mqcFile: mqc_report, mqcMaxSize: params.max_multiqc_email_size.toBytes() ]
    def sf = new File("$baseDir/assets/sendmail_template.txt")
    def sendmail_template = engine.createTemplate(sf).make(smail_fields)
    def sendmail_html = sendmail_template.toString()

    // Send the HTML e-mail
    if (email_address) {
        try {
            if (params.plaintext_email) { throw GroovyException('Send plaintext e-mail, not HTML') }
            // Try to send HTML e-mail using sendmail
            [ 'sendmail', '-t' ].execute() << sendmail_html
            log.info "[nf-core/dualrnaseq] Sent summary e-mail to $email_address (sendmail)"
        } catch (all) {
            // Catch failures and try with plaintext
            [ 'mail', '-s', subject, email_address ].execute() << email_txt
            log.info "[nf-core/dualrnaseq] Sent summary e-mail to $email_address (mail)"
        }
    }

    // Write summary e-mail HTML to a file
    def output_d = new File("${params.outdir}/pipeline_info/")
    if (!output_d.exists()) {
        output_d.mkdirs()
    }
    def output_hf = new File(output_d, "pipeline_report.html")
    output_hf.withWriter { w -> w << email_html }
    def output_tf = new File(output_d, "pipeline_report.txt")
    output_tf.withWriter { w -> w << email_txt }

    c_green = params.monochrome_logs ? '' : "\033[0;32m";
    c_purple = params.monochrome_logs ? '' : "\033[0;35m";
    c_red = params.monochrome_logs ? '' : "\033[0;31m";
    c_reset = params.monochrome_logs ? '' : "\033[0m";

    if (workflow.stats.ignoredCount > 0 && workflow.success) {
        log.info "-${c_purple}Warning, pipeline completed, but with errored process(es) ${c_reset}-"
        log.info "-${c_red}Number of ignored errored process(es) : ${workflow.stats.ignoredCount} ${c_reset}-"
        log.info "-${c_green}Number of successfully ran process(es) : ${workflow.stats.succeedCount} ${c_reset}-"
    }

    if (workflow.success) {
        log.info "-${c_purple}[nf-core/dualrnaseq]${c_green} Pipeline completed successfully${c_reset}-"
    } else {
        checkHostname()
        log.info "-${c_purple}[nf-core/dualrnaseq]${c_red} Pipeline completed with errors${c_reset}-"
    }

}


def nfcoreHeader() {
    // Log colors ANSI codes
    c_black = params.monochrome_logs ? '' : "\033[0;30m";
    c_blue = params.monochrome_logs ? '' : "\033[0;34m";
    c_cyan = params.monochrome_logs ? '' : "\033[0;36m";
    c_dim = params.monochrome_logs ? '' : "\033[2m";
    c_green = params.monochrome_logs ? '' : "\033[0;32m";
    c_purple = params.monochrome_logs ? '' : "\033[0;35m";
    c_reset = params.monochrome_logs ? '' : "\033[0m";
    c_white = params.monochrome_logs ? '' : "\033[0;37m";
    c_yellow = params.monochrome_logs ? '' : "\033[0;33m";

    return """    -${c_dim}--------------------------------------------------${c_reset}-
                                            ${c_green},--.${c_black}/${c_green},-.${c_reset}
    ${c_blue}        ___     __   __   __   ___     ${c_green}/,-._.--~\'${c_reset}
    ${c_blue}  |\\ | |__  __ /  ` /  \\ |__) |__         ${c_yellow}}  {${c_reset}
    ${c_blue}  | \\| |       \\__, \\__/ |  \\ |___     ${c_green}\\`-._,-`-,${c_reset}
                                            ${c_green}`._,._,\'${c_reset}
    ${c_purple}  nf-core/dualrnaseq v${workflow.manifest.version}${c_reset}
    -${c_dim}--------------------------------------------------${c_reset}-
    """.stripIndent()
}

def checkHostname() {
    def c_reset = params.monochrome_logs ? '' : "\033[0m"
    def c_white = params.monochrome_logs ? '' : "\033[0;37m"
    def c_red = params.monochrome_logs ? '' : "\033[1;91m"
    def c_yellow_bold = params.monochrome_logs ? '' : "\033[1;93m"
    if (params.hostnames) {
        def hostname = "hostname".execute().text.trim()
        params.hostnames.each { prof, hnames ->
            hnames.each { hname ->
                if (hostname.contains(hname) && !workflow.profile.contains(prof)) {
                    log.error "====================================================\n" +
                            "  ${c_red}WARNING!${c_reset} You are running with `-profile $workflow.profile`\n" +
                            "  but your machine hostname is ${c_white}'$hostname'${c_reset}\n" +
                            "  ${c_yellow_bold}It's highly recommended that you use `-profile $prof${c_reset}`\n" +
                            "============================================================"
                }
            }
        }
    }
}
